{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "isolated-harvard",
   "metadata": {},
   "source": [
    "# Install\n",
    "\n",
    "To install, after creating conda environment from `environment.yml` run in R:\n",
    "\n",
    "```R\n",
    "install.packages(\"jackalope\")\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "identical-concord",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "< Set of 10 chromosomes >\n",
       "# Total size: 10,000 bp\n",
       "  name                             chromosome                             length\n",
       "chrom0     GACTTCTCCCCTTTCCGACACAAATCTG...TGCTATTCAATCAGCACGAGCTATAAAC      1000\n",
       "chrom1     GCATGCTATGCCGTATAGGTTGTTGCGT...GTCATACCCAATCTAATCGGCCACGAAC      1000\n",
       "chrom2     GCTCTATAAACTCGGGCCGGATCCACGG...TGGCCGTCAGTGGAGAGCGCCGAAAAGA      1000\n",
       "chrom3     TTATTCGTTAACCATGGCGAGATGCACA...CCGTAGCTATCTTAACGTTAGCAGTGGC      1000\n",
       "chrom4     CCGGTAATATCAAGCCGAAAGGCCGGCA...CAGCCCCTACGCACTGAACTAGTGGTCC      1000\n",
       "chrom5     GGGGACCCATTTAGTTTAGGAAAGACGC...CACCGTCGCTACGCGGATTCCACTGATC      1000\n",
       "chrom6     ACCGATCAAGTGGAGATCCGAGCAATTC...CCGAGAGTAAGACCGAATTCTGCCTGTG      1000\n",
       "chrom7     GGCGACTGGAAACTTAAGTCTCACGCGC...AATATGCTACAGTACCTCCTTCTAGATC      1000\n",
       "chrom8     TGGGTATGAGTTGGTCACTAACTAAACC...GTGTTGCGTACCAACCTAGAGGCCGGAC      1000\n",
       "chrom9     AGTCTGTATATCAGCGTCCGAAACGGCG...CGAGCTCAAGGACCGATACGGCGTCGAT      1000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(jackalope)\n",
    "\n",
    "reference <- create_genome(n_chroms = 10, len_mean = 1000)\n",
    "reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "consecutive-certificate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Phylogenetic tree with 5 tips and 4 internal nodes.\n",
       "\n",
       "Tip labels:\n",
       "  t5, t2, t1, t4, t3\n",
       "\n",
       "Rooted; includes branch lengths."
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tr <- ape::rcoal(5)\n",
    "tr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "sharp-ivory",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                              << haplotypes object >>\n",
       "# Haplotypes: 5\n",
       "# Mutations: 16,243\n",
       "\n",
       "                          << Reference genome info: >>\n",
       "< Set of 10 chromosomes >\n",
       "# Total size: 10,000 bp\n",
       "  name                             chromosome                             length\n",
       "chrom0     GACTTCTCCCCTTTCCGACACAAATCTG...TGCTATTCAATCAGCACGAGCTATAAAC      1000\n",
       "chrom1     GCATGCTATGCCGTATAGGTTGTTGCGT...GTCATACCCAATCTAATCGGCCACGAAC      1000\n",
       "chrom2     GCTCTATAAACTCGGGCCGGATCCACGG...TGGCCGTCAGTGGAGAGCGCCGAAAAGA      1000\n",
       "chrom3     TTATTCGTTAACCATGGCGAGATGCACA...CCGTAGCTATCTTAACGTTAGCAGTGGC      1000\n",
       "chrom4     CCGGTAATATCAAGCCGAAAGGCCGGCA...CAGCCCCTACGCACTGAACTAGTGGTCC      1000\n",
       "chrom5     GGGGACCCATTTAGTTTAGGAAAGACGC...CACCGTCGCTACGCGGATTCCACTGATC      1000\n",
       "chrom6     ACCGATCAAGTGGAGATCCGAGCAATTC...CCGAGAGTAAGACCGAATTCTGCCTGTG      1000\n",
       "chrom7     GGCGACTGGAAACTTAAGTCTCACGCGC...AATATGCTACAGTACCTCCTTCTAGATC      1000\n",
       "chrom8     TGGGTATGAGTTGGTCACTAACTAAACC...GTGTTGCGTACCAACCTAGAGGCCGGAC      1000\n",
       "chrom9     AGTCTGTATATCAGCGTCCGAAACGGCG...CGAGCTCAAGGACCGATACGGCGTCGAT      1000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ref_haplotypes <- create_haplotypes(reference, haps_phylo(tr), sub_JC69(0.1))\n",
    "ref_haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "impressed-adjustment",
   "metadata": {},
   "outputs": [],
   "source": [
    "illumina(ref_haplotypes, out_prefix = \"simulated\", sep_files=TRUE, n_reads = 500000,\n",
    "         compress=TRUE, comp_method=\"gzip\",\n",
    "         paired = TRUE, read_length = 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "coated-cabin",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<table width=\"100%\" summary=\"page for illumina {jackalope}\"><tr><td>illumina {jackalope}</td><td style=\"text-align: right;\">R Documentation</td></tr></table>\n",
       "\n",
       "<h2>Create and write Illumina reads to FASTQ file(s).</h2>\n",
       "\n",
       "<h3>Description</h3>\n",
       "\n",
       "<p>From either a reference genome or set of variant haplotypes, create Illumina reads\n",
       "from error profiles and write them to FASTQ output file(s).\n",
       "I encourage you to cite the reference below in addition to <code>jackalope</code> if you use\n",
       "this function.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Usage</h3>\n",
       "\n",
       "<pre>\n",
       "illumina(obj,\n",
       "         out_prefix,\n",
       "         n_reads,\n",
       "         read_length,\n",
       "         paired,\n",
       "         frag_mean = 400,\n",
       "         frag_sd = 100,\n",
       "         matepair = FALSE,\n",
       "         seq_sys = NULL,\n",
       "         profile1 = NULL,\n",
       "         profile2 = NULL,\n",
       "         ins_prob1 = 0.00009,\n",
       "         del_prob1 = 0.00011,\n",
       "         ins_prob2 = 0.00015,\n",
       "         del_prob2 = 0.00023,\n",
       "         frag_len_min = NULL,\n",
       "         frag_len_max = NULL,\n",
       "         haplotype_probs = NULL,\n",
       "         barcodes = NULL,\n",
       "         prob_dup = 0.02,\n",
       "         sep_files = FALSE,\n",
       "         compress = FALSE,\n",
       "         comp_method = \"bgzip\",\n",
       "         n_threads = 1L,\n",
       "         read_pool_size = 1000L,\n",
       "         show_progress = FALSE,\n",
       "         overwrite = FALSE)\n",
       "</pre>\n",
       "\n",
       "\n",
       "<h3>Arguments</h3>\n",
       "\n",
       "<table summary=\"R argblock\">\n",
       "<tr valign=\"top\"><td><code>obj</code></td>\n",
       "<td>\n",
       "<p>Sequencing object of class <code>ref_genome</code> or <code>haplotypes</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>out_prefix</code></td>\n",
       "<td>\n",
       "<p>Prefix for the output file(s), including entire path except\n",
       "for the file extension.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>n_reads</code></td>\n",
       "<td>\n",
       "<p>Number of reads you want to create.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>read_length</code></td>\n",
       "<td>\n",
       "<p>Length of reads.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>paired</code></td>\n",
       "<td>\n",
       "<p>Logical for whether to use paired-end reads.\n",
       "This argument is changed to <code>TRUE</code> if <code>matepair</code> is <code>TRUE</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>frag_mean</code></td>\n",
       "<td>\n",
       "<p>Mean of the Gamma distribution that generates fragment sizes.\n",
       "Defaults to <code>400</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>frag_sd</code></td>\n",
       "<td>\n",
       "<p>Standard deviation of the Gamma distribution that generates\n",
       "fragment sizes.\n",
       "Defaults to <code>100</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>matepair</code></td>\n",
       "<td>\n",
       "<p>Logical for whether to simulate mate-pair reads.\n",
       "Defaults to <code>FALSE</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>seq_sys</code></td>\n",
       "<td>\n",
       "<p>Full or abbreviated name of sequencing system to use.\n",
       "See &quot;Sequencing systems&quot; section for options.\n",
       "See &quot;Sequencing profiles&quot; section for more information on how this argument,\n",
       "<code>profile1</code>, and <code>profile2</code> are used to specify profiles.\n",
       "Defaults to <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>profile1</code></td>\n",
       "<td>\n",
       "<p>Custom profile file for read 1.\n",
       "See &quot;Sequencing profiles&quot; section for more information on how this argument,\n",
       "<code>profile2</code>, and <code>seq_sys</code> are used to specify profiles.\n",
       "Defaults to <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>profile2</code></td>\n",
       "<td>\n",
       "<p>Custom profile file for read 2.\n",
       "See &quot;Sequencing profiles&quot; section for more information on how this argument,\n",
       "<code>profile1</code>, and <code>seq_sys</code> are used to specify profiles.\n",
       "Defaults to <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ins_prob1</code></td>\n",
       "<td>\n",
       "<p>Insertion probability for read 1. Defaults to <code>0.00009</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>del_prob1</code></td>\n",
       "<td>\n",
       "<p>Deletion probability for read 1. Defaults to <code>0.00011</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>ins_prob2</code></td>\n",
       "<td>\n",
       "<p>Insertion probability for read 2. Defaults to <code>0.00015</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>del_prob2</code></td>\n",
       "<td>\n",
       "<p>Deletion probability for read 2. Defaults to <code>0.00023</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>frag_len_min</code></td>\n",
       "<td>\n",
       "<p>Minimum fragment size. A <code>NULL</code> value results in the read length.\n",
       "Defaults to <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>frag_len_max</code></td>\n",
       "<td>\n",
       "<p>Maximum fragment size.\n",
       "A <code>NULL</code> value results in <code>2^32-1</code>, the maximum allowed value.\n",
       "Defaults to <code>NULL</code></p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>haplotype_probs</code></td>\n",
       "<td>\n",
       "<p>Relative probability of sampling each haplotype.\n",
       "This is ignored if sequencing a reference genome.\n",
       "<code>NULL</code> results in all having the same probability.\n",
       "Defaults to <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>barcodes</code></td>\n",
       "<td>\n",
       "<p>Character vector of barcodes for each haplotype, or a single barcode\n",
       "if sequencing a reference genome. <code>NULL</code> results in no barcodes.\n",
       "Defaults to <code>NULL</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>prob_dup</code></td>\n",
       "<td>\n",
       "<p>A single number indicating the probability of duplicates.\n",
       "Defaults to <code>0.02</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>sep_files</code></td>\n",
       "<td>\n",
       "<p>Logical indicating whether to make separate files for each haplotype.\n",
       "This argument is coerced to <code>FALSE</code> if the <code>obj</code> argument is not\n",
       "a <code>haplotypes</code> object.\n",
       "Defaults to <code>FALSE</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>compress</code></td>\n",
       "<td>\n",
       "<p>Logical specifying whether or not to compress output file, or\n",
       "an integer specifying the level of compression, from 1 to 9.\n",
       "If <code>TRUE</code>, a compression level of <code>6</code> is used.\n",
       "Defaults to <code>FALSE</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>comp_method</code></td>\n",
       "<td>\n",
       "<p>Character specifying which type of compression to use if any\n",
       "is desired. Options include <code>\"gzip\"</code> and <code>\"bgzip\"</code>.\n",
       "This is ignored if <code>compress</code> is <code>FALSE</code>, and it throws an error if\n",
       "it's set to <code>\"gzip\"</code> when <code>n_threads &gt; 1</code> (since I don't have a method to\n",
       "do gzip compression in parallel).\n",
       "Defaults to <code>\"bgzip\"</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>n_threads</code></td>\n",
       "<td>\n",
       "<p>The number of threads to use in processing.\n",
       "If <code>compress</code> is <code>TRUE</code> or <code style=\"white-space: pre;\">&gt; 0</code> (indicating compressed output),\n",
       "setting <code>n_threads</code> to <code>2</code> or more makes this function first create an\n",
       "uncompressed file/files using <code>n_threads</code> threads, then compress that/those\n",
       "file/files also using <code>n_threads</code> threads.\n",
       "There is no speed increase if you try to use multiple threads to create\n",
       "compressed output on the fly, so that option is not included.\n",
       "If you want to be conservative with disk space (by not having an uncompressed\n",
       "file present even temporarily), set <code>n_threads</code> to <code>1</code>.\n",
       "Threads are NOT spread across chromosomes or haplotypes, so you don't need to\n",
       "think about these when choosing this argument's value.\n",
       "However, all threads write to the same file/files, so there are diminishing\n",
       "returns for providing many threads.\n",
       "This argument is ignored if the package was not compiled with OpenMP.\n",
       "Defaults to <code>1</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>read_pool_size</code></td>\n",
       "<td>\n",
       "<p>The number of reads to store before writing to disk.\n",
       "Increasing this number should improve speed but take up more memory.\n",
       "Defaults to <code>1000</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>show_progress</code></td>\n",
       "<td>\n",
       "<p>Logical for whether to show a progress bar.\n",
       "Defaults to <code>FALSE</code>.</p>\n",
       "</td></tr>\n",
       "<tr valign=\"top\"><td><code>overwrite</code></td>\n",
       "<td>\n",
       "<p>Logical for whether to overwrite existing FASTQ file(s) of the\n",
       "same name, if they exist.</p>\n",
       "</td></tr>\n",
       "</table>\n",
       "\n",
       "\n",
       "<h3>Value</h3>\n",
       "\n",
       "<p>Nothing is returned.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Sequencing profiles</h3>\n",
       "\n",
       "<p>This section outlines how to use the <code>seq_sys</code>, <code>profile1</code>,\n",
       "and <code>profile2</code> arguments.\n",
       "If all arguments are <code>NULL</code> (their defaults), a sequencing system is chosen\n",
       "based on the read length.\n",
       "If, however, one or more arguments has been provided, then how they're provided\n",
       "should depend on whether you want single- or paired-end reads.\n",
       "</p>\n",
       "<p><strong>For single-end reads</strong>\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li> <p><code>profile2</code> should be <code>NULL</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> Only <code>seq_sys</code> or <code>profile1</code> should be provided, not both.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "<p><strong>For paired-end reads</strong>\n",
       "</p>\n",
       "\n",
       "<ul>\n",
       "<li><p> If providing <code>seq_sys</code>, don't provide either <code>profile1</code> or <code>profile2</code>.\n",
       "</p>\n",
       "</li>\n",
       "<li><p> If providing <code>profile1</code>, you must also provide <code>profile2</code> (they can be the\n",
       "same if you want) and you cannot provide <code>seq_sys</code>.\n",
       "</p>\n",
       "</li></ul>\n",
       "\n",
       "\n",
       "\n",
       "<h3>Sequencing systems</h3>\n",
       "\n",
       "<p>Sequencing system options are the following, where, for each system,\n",
       "&quot;name&quot; is the full name, &quot;abbrev&quot; is the abbreviated name,\n",
       "&quot;max_len&quot; indicates the maximum allowed read length,\n",
       "and\n",
       "&quot;paired&quot; indicates whether paired-end sequencing is allowed.\n",
       "</p>\n",
       "\n",
       "<table summary=\"Rd table\">\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "name                 </td><td style=\"text-align: left;\"> abbrev   </td><td style=\"text-align: left;\"> max_len </td><td style=\"text-align: left;\"> paired </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "Genome Analyzer I    </td><td style=\"text-align: left;\"> GA1      </td><td style=\"text-align: left;\"> 44      </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "Genome Analyzer II   </td><td style=\"text-align: left;\"> GA2      </td><td style=\"text-align: left;\"> 75      </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "HiSeq 1000           </td><td style=\"text-align: left;\"> HS10     </td><td style=\"text-align: left;\"> 100     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "HiSeq 2000           </td><td style=\"text-align: left;\"> HS20     </td><td style=\"text-align: left;\"> 100     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "HiSeq 2500           </td><td style=\"text-align: left;\"> HS25     </td><td style=\"text-align: left;\"> 150     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "HiSeqX v2.5 PCR free </td><td style=\"text-align: left;\"> HSXn     </td><td style=\"text-align: left;\"> 150     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "HiSeqX v2.5 TruSeq   </td><td style=\"text-align: left;\"> HSXt     </td><td style=\"text-align: left;\"> 150     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "MiniSeq TruSeq       </td><td style=\"text-align: left;\"> MinS     </td><td style=\"text-align: left;\"> 50      </td><td style=\"text-align: left;\"> No    </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "MiSeq v1             </td><td style=\"text-align: left;\"> MSv1     </td><td style=\"text-align: left;\"> 250     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "MiSeq v3             </td><td style=\"text-align: left;\"> MSv3     </td><td style=\"text-align: left;\"> 250     </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "NextSeq 500 v2       </td><td style=\"text-align: left;\"> NS50     </td><td style=\"text-align: left;\"> 75      </td><td style=\"text-align: left;\"> Yes   </td>\n",
       "</tr>\n",
       "<tr>\n",
       " <td style=\"text-align: left;\">\n",
       "</td>\n",
       "</tr>\n",
       "\n",
       "</table>\n",
       "\n",
       "\n",
       "\n",
       "<h3>ID lines</h3>\n",
       "\n",
       "<p>The ID lines for FASTQ files are formatted as such:\n",
       "</p>\n",
       "<p><code style=\"white-space: pre;\">@&lt;genome name&gt;-&lt;chromosome name&gt;-&lt;starting position&gt;-&lt;strand&gt;[/&lt;read#&gt;]</code>\n",
       "</p>\n",
       "<p>where the part in <code style=\"white-space: pre;\">[]</code> is only for paired-end Illumina reads, and where <code style=\"white-space: pre;\">genome name</code>\n",
       "is always <code>REF</code> for reference genomes (as opposed to haplotypes).\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>References</h3>\n",
       "\n",
       "<p>Huang, W., L. Li, J. R. Myers, and G. T. Marth. 2012. ART: a next-generation\n",
       "sequencing read simulator. <em>Bioinformatics</em> <strong>28</strong>:593<e2><80><93>594.\n",
       "</p>\n",
       "\n",
       "\n",
       "<h3>Examples</h3>\n",
       "\n",
       "<pre>\n",
       "\n",
       "rg &lt;- create_genome(10, 100e3, 100)\n",
       "dir &lt;- tempdir(TRUE)\n",
       "illumina(rg, paste0(dir, \"/illumina_reads\"), n_reads = 100,\n",
       "         read_length = 100, paired = FALSE)\n",
       "\n",
       "\n",
       "</pre>\n",
       "\n",
       "<hr /><div style=\"text-align: center;\">[Package <em>jackalope</em> version 1.1.1 ]</div>"
      ],
      "text/latex": [
       "\\inputencoding{utf8}\n",
       "\\HeaderA{illumina}{Create and write Illumina reads to FASTQ file(s).}{illumina}\n",
       "%\n",
       "\\begin{Description}\\relax\n",
       "From either a reference genome or set of variant haplotypes, create Illumina reads\n",
       "from error profiles and write them to FASTQ output file(s).\n",
       "I encourage you to cite the reference below in addition to \\code{jackalope} if you use\n",
       "this function.\n",
       "\\end{Description}\n",
       "%\n",
       "\\begin{Usage}\n",
       "\\begin{verbatim}\n",
       "illumina(obj,\n",
       "         out_prefix,\n",
       "         n_reads,\n",
       "         read_length,\n",
       "         paired,\n",
       "         frag_mean = 400,\n",
       "         frag_sd = 100,\n",
       "         matepair = FALSE,\n",
       "         seq_sys = NULL,\n",
       "         profile1 = NULL,\n",
       "         profile2 = NULL,\n",
       "         ins_prob1 = 0.00009,\n",
       "         del_prob1 = 0.00011,\n",
       "         ins_prob2 = 0.00015,\n",
       "         del_prob2 = 0.00023,\n",
       "         frag_len_min = NULL,\n",
       "         frag_len_max = NULL,\n",
       "         haplotype_probs = NULL,\n",
       "         barcodes = NULL,\n",
       "         prob_dup = 0.02,\n",
       "         sep_files = FALSE,\n",
       "         compress = FALSE,\n",
       "         comp_method = \"bgzip\",\n",
       "         n_threads = 1L,\n",
       "         read_pool_size = 1000L,\n",
       "         show_progress = FALSE,\n",
       "         overwrite = FALSE)\n",
       "\\end{verbatim}\n",
       "\\end{Usage}\n",
       "%\n",
       "\\begin{Arguments}\n",
       "\\begin{ldescription}\n",
       "\\item[\\code{obj}] Sequencing object of class \\code{ref\\_genome} or \\code{haplotypes}.\n",
       "\n",
       "\\item[\\code{out\\_prefix}] Prefix for the output file(s), including entire path except\n",
       "for the file extension.\n",
       "\n",
       "\\item[\\code{n\\_reads}] Number of reads you want to create.\n",
       "\n",
       "\\item[\\code{read\\_length}] Length of reads.\n",
       "\n",
       "\\item[\\code{paired}] Logical for whether to use paired-end reads.\n",
       "This argument is changed to \\code{TRUE} if \\code{matepair} is \\code{TRUE}.\n",
       "\n",
       "\\item[\\code{frag\\_mean}] Mean of the Gamma distribution that generates fragment sizes.\n",
       "Defaults to \\code{400}.\n",
       "\n",
       "\\item[\\code{frag\\_sd}] Standard deviation of the Gamma distribution that generates\n",
       "fragment sizes.\n",
       "Defaults to \\code{100}.\n",
       "\n",
       "\\item[\\code{matepair}] Logical for whether to simulate mate-pair reads.\n",
       "Defaults to \\code{FALSE}.\n",
       "\n",
       "\\item[\\code{seq\\_sys}] Full or abbreviated name of sequencing system to use.\n",
       "See \"Sequencing systems\" section for options.\n",
       "See \"Sequencing profiles\" section for more information on how this argument,\n",
       "\\code{profile1}, and \\code{profile2} are used to specify profiles.\n",
       "Defaults to \\code{NULL}.\n",
       "\n",
       "\\item[\\code{profile1}] Custom profile file for read 1.\n",
       "See \"Sequencing profiles\" section for more information on how this argument,\n",
       "\\code{profile2}, and \\code{seq\\_sys} are used to specify profiles.\n",
       "Defaults to \\code{NULL}.\n",
       "\n",
       "\\item[\\code{profile2}] Custom profile file for read 2.\n",
       "See \"Sequencing profiles\" section for more information on how this argument,\n",
       "\\code{profile1}, and \\code{seq\\_sys} are used to specify profiles.\n",
       "Defaults to \\code{NULL}.\n",
       "\n",
       "\\item[\\code{ins\\_prob1}] Insertion probability for read 1. Defaults to \\code{0.00009}.\n",
       "\n",
       "\\item[\\code{del\\_prob1}] Deletion probability for read 1. Defaults to \\code{0.00011}.\n",
       "\n",
       "\\item[\\code{ins\\_prob2}] Insertion probability for read 2. Defaults to \\code{0.00015}.\n",
       "\n",
       "\\item[\\code{del\\_prob2}] Deletion probability for read 2. Defaults to \\code{0.00023}.\n",
       "\n",
       "\\item[\\code{frag\\_len\\_min}] Minimum fragment size. A \\code{NULL} value results in the read length.\n",
       "Defaults to \\code{NULL}.\n",
       "\n",
       "\\item[\\code{frag\\_len\\_max}] Maximum fragment size.\n",
       "A \\code{NULL} value results in \\code{2\\textasciicircum{}32-1}, the maximum allowed value.\n",
       "Defaults to \\code{NULL}\n",
       "\n",
       "\\item[\\code{haplotype\\_probs}] Relative probability of sampling each haplotype.\n",
       "This is ignored if sequencing a reference genome.\n",
       "\\code{NULL} results in all having the same probability.\n",
       "Defaults to \\code{NULL}.\n",
       "\n",
       "\\item[\\code{barcodes}] Character vector of barcodes for each haplotype, or a single barcode\n",
       "if sequencing a reference genome. \\code{NULL} results in no barcodes.\n",
       "Defaults to \\code{NULL}.\n",
       "\n",
       "\\item[\\code{prob\\_dup}] A single number indicating the probability of duplicates.\n",
       "Defaults to \\code{0.02}.\n",
       "\n",
       "\\item[\\code{sep\\_files}] Logical indicating whether to make separate files for each haplotype.\n",
       "This argument is coerced to \\code{FALSE} if the \\code{obj} argument is not\n",
       "a \\code{haplotypes} object.\n",
       "Defaults to \\code{FALSE}.\n",
       "\n",
       "\\item[\\code{compress}] Logical specifying whether or not to compress output file, or\n",
       "an integer specifying the level of compression, from 1 to 9.\n",
       "If \\code{TRUE}, a compression level of \\code{6} is used.\n",
       "Defaults to \\code{FALSE}.\n",
       "\n",
       "\\item[\\code{comp\\_method}] Character specifying which type of compression to use if any\n",
       "is desired. Options include \\code{\"gzip\"} and \\code{\"bgzip\"}.\n",
       "This is ignored if \\code{compress} is \\code{FALSE}, and it throws an error if\n",
       "it's set to \\code{\"gzip\"} when \\code{n\\_threads > 1} (since I don't have a method to\n",
       "do gzip compression in parallel).\n",
       "Defaults to \\code{\"bgzip\"}.\n",
       "\n",
       "\\item[\\code{n\\_threads}] The number of threads to use in processing.\n",
       "If \\code{compress} is \\code{TRUE} or \\AsIs{> 0} (indicating compressed output),\n",
       "setting \\code{n\\_threads} to \\code{2} or more makes this function first create an\n",
       "uncompressed file/files using \\code{n\\_threads} threads, then compress that/those\n",
       "file/files also using \\code{n\\_threads} threads.\n",
       "There is no speed increase if you try to use multiple threads to create\n",
       "compressed output on the fly, so that option is not included.\n",
       "If you want to be conservative with disk space (by not having an uncompressed\n",
       "file present even temporarily), set \\code{n\\_threads} to \\code{1}.\n",
       "Threads are NOT spread across chromosomes or haplotypes, so you don't need to\n",
       "think about these when choosing this argument's value.\n",
       "However, all threads write to the same file/files, so there are diminishing\n",
       "returns for providing many threads.\n",
       "This argument is ignored if the package was not compiled with OpenMP.\n",
       "Defaults to \\code{1}.\n",
       "\n",
       "\\item[\\code{read\\_pool\\_size}] The number of reads to store before writing to disk.\n",
       "Increasing this number should improve speed but take up more memory.\n",
       "Defaults to \\code{1000}.\n",
       "\n",
       "\\item[\\code{show\\_progress}] Logical for whether to show a progress bar.\n",
       "Defaults to \\code{FALSE}.\n",
       "\n",
       "\\item[\\code{overwrite}] Logical for whether to overwrite existing FASTQ file(s) of the\n",
       "same name, if they exist.\n",
       "\\end{ldescription}\n",
       "\\end{Arguments}\n",
       "%\n",
       "\\begin{Value}\n",
       "Nothing is returned.\n",
       "\\end{Value}\n",
       "%\n",
       "\\begin{Section}{Sequencing profiles}\n",
       "\n",
       "This section outlines how to use the \\code{seq\\_sys}, \\code{profile1},\n",
       "and \\code{profile2} arguments.\n",
       "If all arguments are \\code{NULL} (their defaults), a sequencing system is chosen\n",
       "based on the read length.\n",
       "If, however, one or more arguments has been provided, then how they're provided\n",
       "should depend on whether you want single- or paired-end reads.\n",
       "\n",
       "\\strong{For single-end reads}\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} \\code{profile2} should be \\code{NULL}.\n",
       "\\item{} Only \\code{seq\\_sys} or \\code{profile1} should be provided, not both.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\n",
       "\\strong{For paired-end reads}\n",
       "\\begin{itemize}\n",
       "\n",
       "\\item{} If providing \\code{seq\\_sys}, don't provide either \\code{profile1} or \\code{profile2}.\n",
       "\\item{} If providing \\code{profile1}, you must also provide \\code{profile2} (they can be the\n",
       "same if you want) and you cannot provide \\code{seq\\_sys}.\n",
       "\n",
       "\\end{itemize}\n",
       "\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{Sequencing systems}\n",
       "\n",
       "Sequencing system options are the following, where, for each system,\n",
       "\"name\" is the full name, \"abbrev\" is the abbreviated name,\n",
       "\"max\\_len\" indicates the maximum allowed read length,\n",
       "and\n",
       "\"paired\" indicates whether paired-end sequencing is allowed.\n",
       "\n",
       "\n",
       "\\Tabular{llll}{\n",
       "name                 & abbrev   & max\\_len & paired \\\\{}\n",
       "Genome Analyzer I    & GA1      & 44      & Yes   \\\\{}\n",
       "Genome Analyzer II   & GA2      & 75      & Yes   \\\\{}\n",
       "HiSeq 1000           & HS10     & 100     & Yes   \\\\{}\n",
       "HiSeq 2000           & HS20     & 100     & Yes   \\\\{}\n",
       "HiSeq 2500           & HS25     & 150     & Yes   \\\\{}\n",
       "HiSeqX v2.5 PCR free & HSXn     & 150     & Yes   \\\\{}\n",
       "HiSeqX v2.5 TruSeq   & HSXt     & 150     & Yes   \\\\{}\n",
       "MiniSeq TruSeq       & MinS     & 50      & No    \\\\{}\n",
       "MiSeq v1             & MSv1     & 250     & Yes   \\\\{}\n",
       "MiSeq v3             & MSv3     & 250     & Yes   \\\\{}\n",
       "NextSeq 500 v2       & NS50     & 75      & Yes   \\\\{}\n",
       "}\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{Section}{ID lines}\n",
       "\n",
       "The ID lines for FASTQ files are formatted as such:\n",
       "\n",
       "\\AsIs{@<genome name>-<chromosome name>-<starting position>-<strand>[/<read\\#>]}\n",
       "\n",
       "where the part in \\AsIs{[]} is only for paired-end Illumina reads, and where \\AsIs{genome name}\n",
       "is always \\code{REF} for reference genomes (as opposed to haplotypes).\n",
       "\\end{Section}\n",
       "%\n",
       "\\begin{References}\\relax\n",
       "Huang, W., L. Li, J. R. Myers, and G. T. Marth. 2012. ART: a next-generation\n",
       "sequencing read simulator. \\emph{Bioinformatics} \\strong{28}:593<e2><80><93>594.\n",
       "\\end{References}\n",
       "%\n",
       "\\begin{Examples}\n",
       "\\begin{ExampleCode}\n",
       "\n",
       "rg <- create_genome(10, 100e3, 100)\n",
       "dir <- tempdir(TRUE)\n",
       "illumina(rg, paste0(dir, \"/illumina_reads\"), n_reads = 100,\n",
       "         read_length = 100, paired = FALSE)\n",
       "\n",
       "\n",
       "\\end{ExampleCode}\n",
       "\\end{Examples}"
      ],
      "text/plain": [
       "illumina               package:jackalope               R Documentation\n",
       "\n",
       "_\bC_\br_\be_\ba_\bt_\be _\ba_\bn_\bd _\bw_\br_\bi_\bt_\be _\bI_\bl_\bl_\bu_\bm_\bi_\bn_\ba _\br_\be_\ba_\bd_\bs _\bt_\bo _\bF_\bA_\bS_\bT_\bQ _\bf_\bi_\bl_\be(_\bs).\n",
       "\n",
       "_\bD_\be_\bs_\bc_\br_\bi_\bp_\bt_\bi_\bo_\bn:\n",
       "\n",
       "     From either a reference genome or set of variant haplotypes,\n",
       "     create Illumina reads from error profiles and write them to FASTQ\n",
       "     output file(s). I encourage you to cite the reference below in\n",
       "     addition to 'jackalope' if you use this function.\n",
       "\n",
       "_\bU_\bs_\ba_\bg_\be:\n",
       "\n",
       "     illumina(obj,\n",
       "              out_prefix,\n",
       "              n_reads,\n",
       "              read_length,\n",
       "              paired,\n",
       "              frag_mean = 400,\n",
       "              frag_sd = 100,\n",
       "              matepair = FALSE,\n",
       "              seq_sys = NULL,\n",
       "              profile1 = NULL,\n",
       "              profile2 = NULL,\n",
       "              ins_prob1 = 0.00009,\n",
       "              del_prob1 = 0.00011,\n",
       "              ins_prob2 = 0.00015,\n",
       "              del_prob2 = 0.00023,\n",
       "              frag_len_min = NULL,\n",
       "              frag_len_max = NULL,\n",
       "              haplotype_probs = NULL,\n",
       "              barcodes = NULL,\n",
       "              prob_dup = 0.02,\n",
       "              sep_files = FALSE,\n",
       "              compress = FALSE,\n",
       "              comp_method = \"bgzip\",\n",
       "              n_threads = 1L,\n",
       "              read_pool_size = 1000L,\n",
       "              show_progress = FALSE,\n",
       "              overwrite = FALSE)\n",
       "     \n",
       "_\bA_\br_\bg_\bu_\bm_\be_\bn_\bt_\bs:\n",
       "\n",
       "     obj: Sequencing object of class 'ref_genome' or 'haplotypes'.\n",
       "\n",
       "out_prefix: Prefix for the output file(s), including entire path except\n",
       "          for the file extension.\n",
       "\n",
       " n_reads: Number of reads you want to create.\n",
       "\n",
       "read_length: Length of reads.\n",
       "\n",
       "  paired: Logical for whether to use paired-end reads. This argument is\n",
       "          changed to 'TRUE' if 'matepair' is 'TRUE'.\n",
       "\n",
       "frag_mean: Mean of the Gamma distribution that generates fragment\n",
       "          sizes. Defaults to '400'.\n",
       "\n",
       " frag_sd: Standard deviation of the Gamma distribution that generates\n",
       "          fragment sizes. Defaults to '100'.\n",
       "\n",
       "matepair: Logical for whether to simulate mate-pair reads. Defaults to\n",
       "          'FALSE'.\n",
       "\n",
       " seq_sys: Full or abbreviated name of sequencing system to use. See\n",
       "          \"Sequencing systems\" section for options. See \"Sequencing\n",
       "          profiles\" section for more information on how this argument,\n",
       "          'profile1', and 'profile2' are used to specify profiles.\n",
       "          Defaults to 'NULL'.\n",
       "\n",
       "profile1: Custom profile file for read 1. See \"Sequencing profiles\"\n",
       "          section for more information on how this argument,\n",
       "          'profile2', and 'seq_sys' are used to specify profiles.\n",
       "          Defaults to 'NULL'.\n",
       "\n",
       "profile2: Custom profile file for read 2. See \"Sequencing profiles\"\n",
       "          section for more information on how this argument,\n",
       "          'profile1', and 'seq_sys' are used to specify profiles.\n",
       "          Defaults to 'NULL'.\n",
       "\n",
       "ins_prob1: Insertion probability for read 1. Defaults to '0.00009'.\n",
       "\n",
       "del_prob1: Deletion probability for read 1. Defaults to '0.00011'.\n",
       "\n",
       "ins_prob2: Insertion probability for read 2. Defaults to '0.00015'.\n",
       "\n",
       "del_prob2: Deletion probability for read 2. Defaults to '0.00023'.\n",
       "\n",
       "frag_len_min: Minimum fragment size. A 'NULL' value results in the read\n",
       "          length. Defaults to 'NULL'.\n",
       "\n",
       "frag_len_max: Maximum fragment size. A 'NULL' value results in\n",
       "          '2^32-1', the maximum allowed value. Defaults to 'NULL'\n",
       "\n",
       "haplotype_probs: Relative probability of sampling each haplotype. This\n",
       "          is ignored if sequencing a reference genome. 'NULL' results\n",
       "          in all having the same probability. Defaults to 'NULL'.\n",
       "\n",
       "barcodes: Character vector of barcodes for each haplotype, or a single\n",
       "          barcode if sequencing a reference genome. 'NULL' results in\n",
       "          no barcodes. Defaults to 'NULL'.\n",
       "\n",
       "prob_dup: A single number indicating the probability of duplicates.\n",
       "          Defaults to '0.02'.\n",
       "\n",
       "sep_files: Logical indicating whether to make separate files for each\n",
       "          haplotype. This argument is coerced to 'FALSE' if the 'obj'\n",
       "          argument is not a 'haplotypes' object. Defaults to 'FALSE'.\n",
       "\n",
       "compress: Logical specifying whether or not to compress output file, or\n",
       "          an integer specifying the level of compression, from 1 to 9.\n",
       "          If 'TRUE', a compression level of '6' is used. Defaults to\n",
       "          'FALSE'.\n",
       "\n",
       "comp_method: Character specifying which type of compression to use if\n",
       "          any is desired. Options include '\"gzip\"' and '\"bgzip\"'. This\n",
       "          is ignored if 'compress' is 'FALSE', and it throws an error\n",
       "          if it's set to '\"gzip\"' when 'n_threads > 1' (since I don't\n",
       "          have a method to do gzip compression in parallel). Defaults\n",
       "          to '\"bgzip\"'.\n",
       "\n",
       "n_threads: The number of threads to use in processing. If 'compress' is\n",
       "          'TRUE' or > 0 (indicating compressed output), setting\n",
       "          'n_threads' to '2' or more makes this function first create\n",
       "          an uncompressed file/files using 'n_threads' threads, then\n",
       "          compress that/those file/files also using 'n_threads'\n",
       "          threads. There is no speed increase if you try to use\n",
       "          multiple threads to create compressed output on the fly, so\n",
       "          that option is not included. If you want to be conservative\n",
       "          with disk space (by not having an uncompressed file present\n",
       "          even temporarily), set 'n_threads' to '1'. Threads are NOT\n",
       "          spread across chromosomes or haplotypes, so you don't need to\n",
       "          think about these when choosing this argument's value.\n",
       "          However, all threads write to the same file/files, so there\n",
       "          are diminishing returns for providing many threads. This\n",
       "          argument is ignored if the package was not compiled with\n",
       "          OpenMP. Defaults to '1'.\n",
       "\n",
       "read_pool_size: The number of reads to store before writing to disk.\n",
       "          Increasing this number should improve speed but take up more\n",
       "          memory. Defaults to '1000'.\n",
       "\n",
       "show_progress: Logical for whether to show a progress bar. Defaults to\n",
       "          'FALSE'.\n",
       "\n",
       "overwrite: Logical for whether to overwrite existing FASTQ file(s) of\n",
       "          the same name, if they exist.\n",
       "\n",
       "_\bV_\ba_\bl_\bu_\be:\n",
       "\n",
       "     Nothing is returned.\n",
       "\n",
       "_\bS_\be_\bq_\bu_\be_\bn_\bc_\bi_\bn_\bg _\bp_\br_\bo_\bf_\bi_\bl_\be_\bs:\n",
       "\n",
       "     This section outlines how to use the 'seq_sys', 'profile1', and\n",
       "     'profile2' arguments. If all arguments are 'NULL' (their\n",
       "     defaults), a sequencing system is chosen based on the read length.\n",
       "     If, however, one or more arguments has been provided, then how\n",
       "     they're provided should depend on whether you want single- or\n",
       "     paired-end reads.\n",
       "\n",
       "     *For single-end reads*\n",
       "\n",
       "        * 'profile2' should be 'NULL'.\n",
       "\n",
       "        * Only 'seq_sys' or 'profile1' should be provided, not both.\n",
       "\n",
       "     *For paired-end reads*\n",
       "\n",
       "        * If providing 'seq_sys', don't provide either 'profile1' or\n",
       "          'profile2'.\n",
       "\n",
       "        * If providing 'profile1', you must also provide 'profile2'\n",
       "          (they can be the same if you want) and you cannot provide\n",
       "          'seq_sys'.\n",
       "\n",
       "_\bS_\be_\bq_\bu_\be_\bn_\bc_\bi_\bn_\bg _\bs_\by_\bs_\bt_\be_\bm_\bs:\n",
       "\n",
       "     Sequencing system options are the following, where, for each\n",
       "     system, \"name\" is the full name, \"abbrev\" is the abbreviated name,\n",
       "     \"max_len\" indicates the maximum allowed read length, and \"paired\"\n",
       "     indicates whether paired-end sequencing is allowed.\n",
       "\n",
       "       name                  abbrev  max_len  paired \n",
       "       Genome Analyzer I     GA1     44       Yes    \n",
       "       Genome Analyzer II    GA2     75       Yes    \n",
       "       HiSeq 1000            HS10    100      Yes    \n",
       "       HiSeq 2000            HS20    100      Yes    \n",
       "       HiSeq 2500            HS25    150      Yes    \n",
       "       HiSeqX v2.5 PCR free  HSXn    150      Yes    \n",
       "       HiSeqX v2.5 TruSeq    HSXt    150      Yes    \n",
       "       MiniSeq TruSeq        MinS    50       No     \n",
       "       MiSeq v1              MSv1    250      Yes    \n",
       "       MiSeq v3              MSv3    250      Yes    \n",
       "       NextSeq 500 v2        NS50    75       Yes    \n",
       "      \n",
       "_\bI_\bD _\bl_\bi_\bn_\be_\bs:\n",
       "\n",
       "     The ID lines for FASTQ files are formatted as such:\n",
       "\n",
       "     @<genome name>-<chromosome name>-<starting\n",
       "     position>-<strand>[/<read#>]\n",
       "\n",
       "     where the part in [] is only for paired-end Illumina reads, and\n",
       "     where genome name is always 'REF' for reference genomes (as\n",
       "     opposed to haplotypes).\n",
       "\n",
       "_\bR_\be_\bf_\be_\br_\be_\bn_\bc_\be_\bs:\n",
       "\n",
       "     Huang, W., L. Li, J. R. Myers, and G. T. Marth. 2012. ART: a\n",
       "     next-generation sequencing read simulator. _Bioinformatics_\n",
       "     *28*:593<e2><80><93>594.\n",
       "\n",
       "_\bE_\bx_\ba_\bm_\bp_\bl_\be_\bs:\n",
       "\n",
       "     rg <- create_genome(10, 100e3, 100)\n",
       "     dir <- tempdir(TRUE)\n",
       "     illumina(rg, paste0(dir, \"/illumina_reads\"), n_reads = 100,\n",
       "              read_length = 100, paired = FALSE)\n",
       "     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "?illumina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "remarkable-vinyl",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "binary-brighton",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
