{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "embedded-librarian",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "Commits of thesis-index code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "exposed-reception",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/workspace/thesis-data-simulation/jackalope/salmonella\n",
      "40e716a7bf1a72907bbdf06e3c6f2facb7b97058\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!git -C ../../thesis-index rev-parse HEAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "prostate-university",
   "metadata": {},
   "source": [
    "# Load up variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "informative-space",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-03-15 13:00:45\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,57:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-03-15 13:00:46\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,61:\u001b[0m Use database directory seq_repo\n",
      "Loading phylogeny\n",
      "Loaded variants from [phylogeny] into database\n",
      "Took 4.95 minutes\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import time\n",
    "\n",
    "seq_repo = 'repo'\n",
    "if os.path.exists(seq_repo):\n",
    "    shutil.rmtree(seq_repo)\n",
    "os.mkdir(seq_repo)\n",
    "\n",
    "start_time = time.time()\n",
    "!variants \\\n",
    "    --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --database-dir seq_repo --verbose \\\n",
    "    load-snippy --reference-file input/S_HeidelbergSL476.fasta.gz phylogeny\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dying-depth",
   "metadata": {},
   "source": [
    "Make sure data loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "meaning-turkey",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: variants list [OPTIONS]\n",
      "Try 'variants list --help' for help.\n",
      "\n",
      "Error: Invalid value for '--type': invalid choice: genomes. (choose from genome, sample, reference)\n"
     ]
    }
   ],
   "source": [
    "!variants --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' --database-dir seq_repo \\\n",
    "    list --type genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "contained-hayes",
   "metadata": {},
   "source": [
    "# Build tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "abstract-hormone",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-03-15 13:05:45\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,57:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-03-15 13:05:45\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,61:\u001b[0m Use database directory seq_repo\n",
      "\u001b[32m2021-03-15 13:07:21\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._create_core_mask,45:\u001b[0m \u001b[32mStarted creating core mask for 59 sequences\u001b[0m\n",
      "\u001b[32m2021-03-15 13:07:21\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._create_core_mask,53:\u001b[0m \u001b[32mFinished creating core mask for 59 sequences. Took 0.02 seconds\u001b[0m\n",
      "\u001b[32m2021-03-15 13:07:21\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._build_core_alignment_sequence,64:\u001b[0m \u001b[32mStarted building core alignment for NC_011083.1 using 73084 variant positions\u001b[0m\n",
      "\u001b[32m2021-03-15 13:08:01\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._build_core_alignment_sequence,96:\u001b[0m \u001b[32mFinished building core alignment for NC_011083.1. Took 39.87 seconds\u001b[0m\n",
      "\u001b[32m2021-03-15 13:08:01\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.TreeService.build_tree,77:\u001b[0m \u001b[32mRunning: iqtree --terrace -T 32 -s /tmp/tmpwjlusc44/input.fasta -m MFP+ASC\u001b[0m\n",
      "Wrote tree to [thesis.core.tre]\n",
      "Wrote log file to [thesis.core.tre.log]\n",
      "Took 4.37 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "!python -mcProfile -o profile.tree.data `which variants` \\\n",
    "    --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --database-dir seq_repo --verbose \\\n",
    "    tree --reference-name S_HeidelbergSL476 --output-file thesis.core.tre --align-type core --threads 32\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "mathematical-alert",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(SH08-001:0.000746112,((((SH09-29:7.99266e-05,((((((((((SH14-001:3.42528e-05,SH14-010:4.08071e-05)1:1.531e-07,(((SH14-007:1.4062e-06,SH14-016:3.411e-07)1:3.412e-07,SH14-019:1.853e-07)1:3.412e-07,SH14-014:8.276e-07)1:1.531e-07)1:1.531e-07,(SH14-013:3.411e-07,SH14-020:1.0196e-06)1:1.531e-07)1:1.531e-07,((SH14-018:1.5653e-06,SH14-022:1.1678e-06)1:2.913e-07,SH14-024:3.411e-07)1:1.531e-07)1:1.531e-07,(SH14-006:7.151e-07,SH14-017:4.08077e-05)1:1.531e-07)1:1.531e-07,(((SH14-002:1.5652e-06,SH14-015:4.808e-07)1:3.886e-07,SH14-009:7.673e-07)1:3.412e-07,SH14-027:6.817e-07)1:1.531e-07)1:1.531e-07,((SH14-003:6.684e-07,SH14-026:9.194e-07)1:1.531e-07,SH14-028:1.531e-07)1:1.531e-07)1:1.531e-07,((SH14-004:2.1743e-06,SH14-011:1.9018e-06)1:1.531e-07,((SH14-005:1.0335e-06,SH14-023:1.844e-07)1:3.412e-07,(SH14-008:8.967e-07,SH14-025:1.0077e-06)1:1.531e-07)1:3.411e-07)1:1.531e-07)1:1.531e-07,SH14-021:4.809e-07)1:2.2572e-06,SH14-012:0.00128268)1:0.000166372)1:1.4799e-06,SH10-30:0.000152028)1:0.000737539,((SH10-001:0.000124319,(SH11-001:4.3197e-05,SH12-011:0.000162073)1:3.292e-07)1:1.3564e-06,((((((SH12-001:2.2017e-06,SH12-010:3.72023e-05)1:1.531e-07,(SH12-008:3.41e-07,SH12-009:7.575e-07)1:1.531e-07)1:1.531e-07,(SH12-003:1.1546e-06,SH12-006:1.897e-06)1:1.531e-07)1:1.531e-07,(SH12-002:1.7471e-06,SH12-007:3.8261e-05)1:3.77008e-05)1:3.37e-07,SH12-005:2.0576e-06)1:4.478e-07,SH12-004:4.522e-07)1:0.000192555)1:0.000872964)1:0.000151358,((SH10-014:7.43881e-05,(SH10-015:9.437e-07,(SH11-002:4.18503e-05,((((SH13-001:2.2235e-06,(SH13-004:4.46494e-05,SH13-006:9.194e-07)1:3.412e-07)1:1.531e-07,(SH13-002:6.683e-07,(SH13-003:2.5151e-06,SH13-007:4.809e-07)1:3.003e-07)1:1.531e-07)1:1.531e-07,SH13-005:1.4062e-06)1:2.12e-07,SH13-008:3.58272e-05)1:7.98273e-05)1:4.33933e-05)1:2.3317e-06)1:0.0017107,((SH12-013:0.000707626,reference:3.2e-07)1:0.00112212,SH12-014:0.00109331)1:0.000422634)1:7.3412e-06)1:0.000655984,(SH10-002:0.000569893,SH12-012:0.000540109)1:4.0093e-06);"
     ]
    }
   ],
   "source": [
    "!cat thesis.core.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "isolated-worker",
   "metadata": {},
   "source": [
    "## Profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "cheap-crazy",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Mar 15 13:10:05 2021    profile.tree.data\n",
      "\n",
      "         183636625 function calls (179106849 primitive calls) in 261.240 seconds\n",
      "\n",
      "   Ordered by: internal time, name/file/line\n",
      "   List reduced from 9087 to 10 due to restriction <10>\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "      165  123.901    0.751  123.901    0.751 {method 'poll' of 'select.poll' objects}\n",
      "    74399   33.216    0.000   33.216    0.000 {method 'recv_into' of '_socket.socket' objects}\n",
      "  4267260   11.640    0.000   13.886    0.000 Seq.py:222(__add__)\n",
      "  4267260    7.239    0.000   34.906    0.000 SeqRecord.py:818(__add__)\n",
      "  4340405    6.944    0.000   12.561    0.000 SeqRecord.py:160(__init__)\n",
      "        1    3.612    3.612   39.869   39.869 CoreAlignmentService.py:58(_build_core_alignment_sequence)\n",
      "  1433732    3.200    0.000   45.767    0.000 connections.py:683(_read_packet)\n",
      " 40278896    3.017    0.000    3.018    0.000 {built-in method builtins.isinstance}\n",
      "  1064452    2.502    0.000   10.007    0.000 loading.py:508(_instance)\n",
      "  2941314    2.387    0.000    2.387    0.000 {method 'settimeout' of '_socket.socket' objects}\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pstats.Stats at 0x7f9ac0ed4040>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pstats\n",
    "from pstats import SortKey\n",
    "p_tree = pstats.Stats('profile.tree.data').strip_dirs().sort_stats(SortKey.TIME, SortKey.NFL)\n",
    "# p_tree.print_stats(10)\n",
    "p_tree.print_stats(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unauthorized-conjunction",
   "metadata": {},
   "source": [
    "# Database size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "sudden-hybrid",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table                    Data size (MB)  Index size (MB)\n",
      "reference                0.02            0.00\n",
      "reference_sequence       0.02            0.02\n",
      "sample                   0.02            0.00\n",
      "sample_kmer_index        0.02            0.00\n",
      "sample_sequence          7.52            0.02\n",
      "sample_variation_allele  56.59           77.27\n",
      "variation_allele         10.55           5.52\n",
      "Total Data (MB)   74.74\n",
      "Total Index (MB)  82.83\n",
      "Total (MB)        157.57\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Fix warning about locale unset\n",
    "os.environ['LANG'] = 'en_US.UTF-8'\n",
    "\n",
    "!mysql -utest -ptest < query-db-size.txt | tee /tmp/total-db-sizes.txt | column -s$'\\t' -t\n",
    "!perl -e '$sd=0;$si=0;$s=0; while(<>){chomp; ($n,$ds,$is) = ($_ =~ /^(\\S+)\\s+(\\S+)\\s+(\\S+)/); if ($n ne \"Table\"){$s += $ds + $is; $si+=$is; $sd += $ds;}} print(\"\\nTotal Data (MB)\\t$sd\\nTotal Index (MB)\\t$si\\nTotal (MB)\\t$s\\n\");' < /tmp/total-db-sizes.txt \\\n",
    "    | column -s$'\\t' -t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "national-tissue",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count(*)\n",
      "1017848\n"
     ]
    }
   ],
   "source": [
    "!echo \"select count(*) from sample_variation_allele;\" | mysql -utest -ptest --database thesis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "municipal-clearance",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count(*)\n",
      "76534\n"
     ]
    }
   ],
   "source": [
    "!echo \"select count(*) from variation_allele;\" | mysql -utest -ptest --database thesis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "limiting-biology",
   "metadata": {},
   "source": [
    "# Compare above with size of input files\n",
    "\n",
    "This copies all the necessary input files to a temporary directory, compresses them, and checks size. This is to compare with size of storing information in database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "expensive-architect",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total size\n",
      "111M\t/tmp/tmpwqf2cd09\n",
      "Size of fasta\n",
      "86M\ttotal\n",
      "Size of VCF\n",
      "26M\ttotal\n",
      "Count of VCF\n",
      "59\n"
     ]
    }
   ],
   "source": [
    "import tempfile\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "with tempfile.TemporaryDirectory() as tmp_dir:\n",
    "    phy_path = os.path.abspath('phylogeny')\n",
    "    phy_names = [d for d in os.listdir(phy_path) if (os.path.isdir(f'{phy_path}/{d}') and d.startswith('SH'))]\n",
    "    fasta_files = []\n",
    "    for name in phy_names:\n",
    "        os.symlink(f'{phy_path}/{name}/snps.vcf.gz', f'{tmp_dir}/{name}.snps.vcf.gz')\n",
    "        shutil.copyfile(f'{phy_path}/{name}/snps.aligned.fa', f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "        fasta_files.append(f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "    !parallel -j 32 -I% gzip % ::: {' '.join(fasta_files)}\n",
    "    print('Total size')\n",
    "    !du -Lsh {tmp_dir}\n",
    "    print('Size of fasta')\n",
    "    !du -Lch {tmp_dir}/*.snps.aligned.fa.gz | tail -n 1\n",
    "    print('Size of VCF')\n",
    "    !du -Lch {tmp_dir}/*.snps.vcf.gz | tail -n 1\n",
    "    print('Count of VCF')\n",
    "    !ls {tmp_dir}/*.snps.vcf.gz | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "industrial-couple",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
