{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "identical-editor",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "Commits of thesis-index code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "printable-canvas",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/workspace/thesis-data-simulation/jackalope/salmonella\n",
      "bb29ff5937af4bd390263a0f9ff2fbacbb91ca00\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!git -C ../../thesis-index rev-parse HEAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "female-worst",
   "metadata": {},
   "source": [
    "# Load up variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "reserved-score",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-02-15 18:22:21\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,49:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-02-15 18:22:22\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,52:\u001b[0m Use seqrepo directory seq_repo\n",
      "Loading phylogeny\n",
      "Loaded variants from [phylogeny] into database\n",
      "Took 9.65 minutes\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import time\n",
    "\n",
    "seq_repo = 'repo'\n",
    "if os.path.exists(seq_repo):\n",
    "    shutil.rmtree(seq_repo)\n",
    "os.mkdir(seq_repo)\n",
    "\n",
    "start_time = time.time()\n",
    "!python -mcProfile -o profile.load.data `which variants` \\\n",
    "    --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --seqrepo-dir seq_repo --verbose \\\n",
    "    load --reference-file input/S_HeidelbergSL476.fasta.gz phylogeny\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "changing-groove",
   "metadata": {},
   "source": [
    "Make sure data loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "unsigned-structure",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_HeidelbergSL476\n"
     ]
    }
   ],
   "source": [
    "!variants --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' --seqrepo-dir seq_repo \\\n",
    "    list --type genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "computational-seventh",
   "metadata": {},
   "source": [
    "## Look at profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "median-junction",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Feb 15 18:31:59 2021    profile.load.data\n",
      "\n",
      "         1439226134 function calls (1431015737 primitive calls) in 577.457 seconds\n",
      "\n",
      "   Ordered by: internal time, name/file/line\n",
      "   List reduced from 9178 to 20 due to restriction <20>\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "       59  100.125    1.697  186.737    3.165 CoreBitMask.py:40(from_sequence)\n",
      "288437371   51.671    0.000   67.275    0.000 Seq.py:208(__getitem__)\n",
      "  1017848   40.111    0.000   48.079    0.000 parser.py:436(_parse_samples)\n",
      "355555437   28.975    0.000   35.848    0.000 {built-in method builtins.isinstance}\n",
      "288438396   19.346    0.000   19.346    0.000 {method 'upper' of 'str' objects}\n",
      "      157   16.930    0.108   16.930    0.108 {method 'recv_into' of '_socket.socket' objects}\n",
      "  1017848    8.928    0.000   20.322    0.000 parser.py:362(_parse_info)\n",
      "  1020385    8.207    0.000   10.380    0.000 {pandas._libs.lib.infer_dtype}\n",
      "  5395376    8.018    0.000   41.558    0.000 series.py:837(__getitem__)\n",
      "  9160632    7.883    0.000   15.384    0.000 parser.py:357(_map)\n",
      "  1018439    6.940    0.000  120.721    0.000 series.py:236(__init__)\n",
      "  1017907    6.625    0.000  106.911    0.000 parser.py:540(__next__)\n",
      "  2037124    5.840    0.000    5.840    0.000 {method 'split' of 're.Pattern' objects}\n",
      "        1    5.751    5.751  183.158  183.158 VariationService.py:43(_create_file_variants)\n",
      "  1019382    5.319    0.000   60.636    0.000 construction.py:423(sanitize_array)\n",
      " 20475199    5.166    0.000    6.871    0.000 generic.py:30(_check)\n",
      "  1019500    5.124    0.000   14.412    0.000 codecs.py:531(readline)\n",
      "  5395376    5.110    0.000   28.949    0.000 series.py:942(_get_value)\n",
      "    76653    4.212    0.000    4.212    0.000 attributes.py:1750(<listcomp>)\n",
      "  1020798    4.188    0.000   17.309    0.000 cast.py:1379(maybe_cast_to_datetime)\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pstats.Stats at 0x7f7c0063c070>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pstats\n",
    "from pstats import SortKey\n",
    "p_load = pstats.Stats('profile.load.data').strip_dirs().sort_stats(SortKey.TIME, SortKey.NFL)\n",
    "p_load.print_stats(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "attached-commerce",
   "metadata": {},
   "source": [
    "# Build tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "coupled-basics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-02-15 18:33:06\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,49:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-02-15 18:33:06\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,52:\u001b[0m Use seqrepo directory seq_repo\n",
      "\u001b[32m2021-02-15 18:34:40\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._create_core_mask,46:\u001b[0m \u001b[32mStarted creating core mask for 59 sequences\u001b[0m\n",
      "\u001b[32m2021-02-15 18:34:40\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._create_core_mask,54:\u001b[0m \u001b[32mFinished creating core mask for 59 sequences. Took 0.02 seconds\u001b[0m\n",
      "\u001b[32m2021-02-15 18:34:40\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._build_core_alignment_sequence,65:\u001b[0m \u001b[32mStarted building core alignment for NC_011083.1 using 73084 variant positions\u001b[0m\n",
      "\u001b[32m2021-02-15 18:35:20\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._build_core_alignment_sequence,97:\u001b[0m \u001b[32mFinished building core alignment for NC_011083.1. Took 40.22 seconds\u001b[0m\n",
      "\u001b[32m2021-02-15 18:35:21\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.TreeService.build_tree,77:\u001b[0m \u001b[32mRunning: iqtree --terrace -T 32 -s /tmp/tmpp6vvevfw/input.fasta -m MFP+ASC\u001b[0m\n",
      "Wrote tree to [thesis.core.tre]\n",
      "Wrote log file to [thesis.core.tre.log]\n",
      "Took 4.60 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "!python -mcProfile -o profile.tree.data `which variants` \\\n",
    "    --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --seqrepo-dir seq_repo --verbose \\\n",
    "    tree --reference-name S_HeidelbergSL476 --output-file thesis.core.tre --align-type core --threads 32\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "incomplete-physiology",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(SH08-001:0.00257412,((((SH09-29:0.000275846,(((((((((SH14-001:0.000118712,SH14-017:0.000140868)1:4.6e-09,SH14-021:1.706e-06)1:4.6e-09,((SH14-003:2.2728e-06,SH14-027:2.84e-06)1:4.6e-09,((SH14-004:6.8186e-06,SH14-024:4.6e-09)1:1.14e-06,(SH14-006:2.8399e-06,SH14-020:3.9774e-06)1:4.6e-09)1:4.6e-09)1:4.6e-09)1:4.6e-09,(((((SH14-002:5.6823e-06,SH14-015:1.7057e-06)1:5.668e-07,SH14-009:2.84e-06)1:1.14e-06,SH14-011:6.8187e-06)1:4.6e-09,(SH14-005:3.9774e-06,SH14-013:1.1401e-06)1:4.6e-09)1:4.6e-09,(SH14-010:0.000140868,SH14-026:3.408e-06)1:4.6e-09)1:4.6e-09)1:4.6e-09,((SH14-008:3.408e-06,(SH14-023:1.1401e-06,SH14-025:3.9774e-06)1:4.6e-09)1:1.14e-06,(SH14-018:5.6822e-06,SH14-022:4.5458e-06)1:5.671e-07)1:4.6e-09)1:4.6e-09,SH14-028:4.6e-09)1:1.22e-08,((SH14-007:5.1139e-06,SH14-016:1.1401e-06)1:1.27e-08,SH14-019:1.14e-06)1:1.14e-06)1:1.047e-06,SH14-014:2.3642e-06)1:7.7531e-06,SH14-012:0.00442831)1:0.00057366)1:5.4746e-06,SH10-30:0.00052467)1:0.00254588,((SH10-001:0.000429041,(SH11-001:0.00014941,SH12-011:0.000559327)1:1.0659e-06)1:4.7109e-06,(SH12-001:6.7669e-06,((((((SH12-002:6.215e-06,SH12-007:0.000132377)1:0.000130104,SH12-010:0.000128367)1:4.6e-09,SH12-009:2.8423e-06)1:4.6e-09,(SH12-003:4.5476e-06,(SH12-004:3.4107e-06,SH12-005:7.3841e-06)1:1.136e-06)1:4.6e-09)1:4.6e-09,SH12-006:6.8159e-06)1:4.6e-09,SH12-008:1.1361e-06)1:1.1857e-06)1:0.000664255)1:0.00301333)1:0.00052194,((SH10-014:0.000256813,(SH10-015:3.8028e-06,(SH11-002:0.000144274,((((SH13-001:7.9541e-06,(SH13-004:0.000153923,SH13-006:3.4104e-06)1:1.1345e-06)1:4.6e-09,((SH13-003:9.0905e-06,SH13-007:1.7014e-06)1:1.136e-06,SH13-005:5.1133e-06)1:4.6e-09)1:4.6e-09,SH13-002:2.2684e-06)1:5.465e-07,SH13-008:0.000123835)1:0.000275563)1:0.000149566)1:8.0861e-06)1:0.00589941,((SH12-013:0.00244036,reference:4.6e-09)1:0.00387164,SH12-014:0.00377262)1:0.00145729)1:2.56416e-05)1:0.00226358,(SH10-002:0.0019661,SH12-012:0.00186317)1:1.44246e-05);"
     ]
    }
   ],
   "source": [
    "!cat thesis.core.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "exterior-domestic",
   "metadata": {},
   "source": [
    "## Profiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "efficient-heating",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Feb 15 18:37:40 2021    profile.tree.data\n",
      "\n",
      "         181863505 function calls (177336739 primitive calls) in 275.111 seconds\n",
      "\n",
      "   Ordered by: internal time, name/file/line\n",
      "   List reduced from 8297 to 10 due to restriction <10>\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "      150  139.734    0.932  139.734    0.932 {method 'poll' of 'select.poll' objects}\n",
      "    74545   31.512    0.000   31.512    0.000 {method 'recv_into' of '_socket.socket' objects}\n",
      "  4267260   12.044    0.000   14.351    0.000 Seq.py:222(__add__)\n",
      "  4267260    7.200    0.000   35.111    0.000 SeqRecord.py:818(__add__)\n",
      "  4340405    6.842    0.000   12.351    0.000 SeqRecord.py:160(__init__)\n",
      "        1    3.716    3.716   40.218   40.218 CoreAlignmentService.py:59(_build_core_alignment_sequence)\n",
      "  1433721    3.267    0.000   44.366    0.000 connections.py:683(_read_packet)\n",
      " 40194112    3.018    0.000    3.019    0.000 {built-in method builtins.isinstance}\n",
      "  1064452    2.485    0.000    9.322    0.000 loading.py:508(_instance)\n",
      "  2941291    2.401    0.000    2.401    0.000 {method 'settimeout' of '_socket.socket' objects}\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pstats.Stats at 0x7f7c0063c3d0>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pstats\n",
    "from pstats import SortKey\n",
    "p_tree = pstats.Stats('profile.tree.data').strip_dirs().sort_stats(SortKey.TIME, SortKey.NFL)\n",
    "# p_tree.print_stats(10)\n",
    "p_tree.print_stats(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "helpful-forestry",
   "metadata": {},
   "source": [
    "# Database size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "second-transcription",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table                    Data size (MB)  Index size (MB)\n",
      "reference                0.02            0.00\n",
      "reference_sequence       0.02            0.02\n",
      "sample                   0.02            0.00\n",
      "sample_sequence          7.52            0.02\n",
      "sample_variation_allele  53.58           71.19\n",
      "variation_allele         10.55           5.52\n",
      "Total Data (MB)   71.71\n",
      "Total Index (MB)  76.75\n",
      "Total (MB)        148.46\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Fix warning about locale unset\n",
    "os.environ['LANG'] = 'en_US.UTF-8'\n",
    "\n",
    "!mysql -utest -ptest < query-db-size.txt | tee /tmp/total-db-sizes.txt | column -s$'\\t' -t\n",
    "!perl -e '$sd=0;$si=0;$s=0; while(<>){chomp; ($n,$ds,$is) = ($_ =~ /^(\\S+)\\s+(\\S+)\\s+(\\S+)/); if ($n ne \"Table\"){$s += $ds + $is; $si+=$is; $sd += $ds;}} print(\"\\nTotal Data (MB)\\t$sd\\nTotal Index (MB)\\t$si\\nTotal (MB)\\t$s\\n\");' < /tmp/total-db-sizes.txt \\\n",
    "    | column -s$'\\t' -t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blocked-junction",
   "metadata": {},
   "source": [
    "# Compare above with size of input files\n",
    "\n",
    "This copies all the necessary input files to a temporary directory, compresses them, and checks size. This is to compare with size of storing information in database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "cosmetic-dover",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total size\n",
      "111M\t/tmp/tmprql03p2l\n",
      "Size of fasta\n",
      "86M\ttotal\n",
      "Size of VCF\n",
      "26M\ttotal\n",
      "Count of VCF\n",
      "59\n"
     ]
    }
   ],
   "source": [
    "import tempfile\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "with tempfile.TemporaryDirectory() as tmp_dir:\n",
    "    phy_path = os.path.abspath('phylogeny')\n",
    "    phy_names = [d for d in os.listdir(phy_path) if (os.path.isdir(f'{phy_path}/{d}') and d.startswith('SH'))]\n",
    "    fasta_files = []\n",
    "    for name in phy_names:\n",
    "        os.symlink(f'{phy_path}/{name}/snps.vcf.gz', f'{tmp_dir}/{name}.snps.vcf.gz')\n",
    "        shutil.copyfile(f'{phy_path}/{name}/snps.aligned.fa', f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "        fasta_files.append(f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "    !parallel -j 32 -I% gzip % ::: {' '.join(fasta_files)}\n",
    "    print('Total size')\n",
    "    !du -Lsh {tmp_dir}\n",
    "    print('Size of fasta')\n",
    "    !du -Lch {tmp_dir}/*.snps.aligned.fa.gz | tail -n 1\n",
    "    print('Size of VCF')\n",
    "    !du -Lch {tmp_dir}/*.snps.vcf.gz | tail -n 1\n",
    "    print('Count of VCF')\n",
    "    !ls {tmp_dir}/*.snps.vcf.gz | wc -l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "mature-abuse",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
