{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "coated-expression",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "Commits of thesis-index code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "individual-newfoundland",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/workspace/thesis-data-simulation/jackalope/salmonella\n",
      "fa9dcf188ab756cc730130c05cedbb106340a631\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!git -C ../../thesis-index rev-parse HEAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "loaded-alfred",
   "metadata": {},
   "source": [
    "# Load up variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cleared-hydrogen",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-02-15 11:29:42\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,49:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-02-15 11:29:43\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,52:\u001b[0m Use seqrepo directory seq_repo\n",
      "Loading phylogeny\n",
      "Loaded variants from [phylogeny] into database\n",
      "Took 9.64 minutes\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import time\n",
    "\n",
    "seq_repo = 'repo'\n",
    "if os.path.exists(seq_repo):\n",
    "    shutil.rmtree(seq_repo)\n",
    "os.mkdir(seq_repo)\n",
    "\n",
    "start_time = time.time()\n",
    "!python -mcProfile -o profile.load.data `which variants` \\\n",
    "    --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --seqrepo-dir seq_repo --verbose \\\n",
    "    load --reference-file input/S_HeidelbergSL476.fasta.gz phylogeny\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "emerging-confidentiality",
   "metadata": {},
   "source": [
    "Make sure data loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "every-development",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "S_HeidelbergSL476\n"
     ]
    }
   ],
   "source": [
    "!variants --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' --seqrepo-dir seq_repo \\\n",
    "    list --type genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mighty-prophet",
   "metadata": {},
   "source": [
    "## Look at profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "institutional-miami",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mon Feb 15 11:39:19 2021    profile.load.data\n",
      "\n",
      "         1440518558 function calls (1432307870 primitive calls) in 576.634 seconds\n",
      "\n",
      "   Ordered by: internal time, name/file/line\n",
      "   List reduced from 9287 to 20 due to restriction <20>\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "       59   95.646    1.621  182.871    3.100 CoreBitMask.py:29(_add_sequence)\n",
      "288437371   52.422    0.000   67.865    0.000 Seq.py:208(__getitem__)\n",
      "  1017848   40.239    0.000   48.228    0.000 parser.py:436(_parse_samples)\n",
      "355810884   28.562    0.000   35.283    0.000 {built-in method builtins.isinstance}\n",
      "288438400   19.394    0.000   19.394    0.000 {method 'upper' of 'str' objects}\n",
      "      206   17.939    0.087   17.939    0.087 {method 'recv_into' of '_socket.socket' objects}\n",
      "  1017848    8.710    0.000   19.977    0.000 parser.py:362(_parse_info)\n",
      "  5395376    8.138    0.000   42.122    0.000 series.py:837(__getitem__)\n",
      "  9160632    7.954    0.000   15.305    0.000 parser.py:357(_map)\n",
      "  1020385    7.721    0.000    9.815    0.000 {pandas._libs.lib.infer_dtype}\n",
      "  1018439    6.851    0.000  118.798    0.000 series.py:236(__init__)\n",
      "  1017907    6.667    0.000  106.375    0.000 parser.py:540(__next__)\n",
      "  2037115    5.844    0.000    5.844    0.000 {method 'split' of 're.Pattern' objects}\n",
      "        1    5.781    5.781  181.619  181.619 VariationService.py:43(_create_file_variants)\n",
      "  1019382    5.253    0.000   59.410    0.000 construction.py:423(sanitize_array)\n",
      "  5395376    5.132    0.000   29.425    0.000 series.py:942(_get_value)\n",
      " 20475199    5.041    0.000    6.718    0.000 generic.py:30(_check)\n",
      "  1019500    4.922    0.000   14.083    0.000 codecs.py:531(readline)\n",
      "27075847/24009544    3.966    0.000    4.938    0.000 {built-in method builtins.len}\n",
      "  1020798    3.917    0.000   17.035    0.000 cast.py:1379(maybe_cast_to_datetime)\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pstats.Stats at 0x7f09ff8f5ee0>"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pstats\n",
    "from pstats import SortKey\n",
    "p_load = pstats.Stats('profile.load.data').strip_dirs().sort_stats(SortKey.TIME, SortKey.NFL)\n",
    "p_load.print_stats(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "popular-commission",
   "metadata": {},
   "source": [
    "# Build tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "macro-frost",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-02-15 12:04:33\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,49:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-02-15 12:04:33\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,52:\u001b[0m Use seqrepo directory seq_repo\n",
      "\u001b[32m2021-02-15 12:06:07\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._create_core_mask,46:\u001b[0m \u001b[32mStarted creating core mask for 59 sequences\u001b[0m\n",
      "\u001b[32m2021-02-15 12:06:07\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._create_core_mask,54:\u001b[0m \u001b[32mFinished creating core mask for 59 sequences. Took 0.03 seconds\u001b[0m\n",
      "\u001b[32m2021-02-15 12:06:07\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._build_core_alignment_sequence,65:\u001b[0m \u001b[32mStarted building core alignment for NC_011083.1 using 73084 variant positions\u001b[0m\n",
      "\u001b[32m2021-02-15 12:06:47\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService._build_core_alignment_sequence,97:\u001b[0m \u001b[32mFinished building core alignment for NC_011083.1. Took 39.76 seconds\u001b[0m\n",
      "\u001b[32m2021-02-15 12:06:47\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.TreeService.build_tree,77:\u001b[0m \u001b[32mRunning: iqtree --terrace -T 32 -s /tmp/tmp970uo31v/input.fasta -m MFP+ASC\u001b[0m\n",
      "Wrote tree to [thesis.core.tre]\n",
      "Wrote log file to [thesis.core.tre.log]\n",
      "Took 5.18 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "!python -mcProfile -o profile.tree.data `which variants` \\\n",
    "    --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --seqrepo-dir seq_repo --verbose \\\n",
    "    tree --reference-name S_HeidelbergSL476 --output-file thesis.core.tre --align-type core --threads 32\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "upper-bristol",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(SH08-001:0.00244467,((((SH09-29:0.000261979,((((((((((((SH14-001:0.000112748,SH14-011:6.476e-06)1:4.5e-09,((SH14-004:6.476e-06,SH14-024:4.5e-09)1:1.0821e-06,SH14-026:3.2378e-06)1:4.5e-09)1:4.5e-09,(SH14-003:2.1585e-06,(SH14-005:2.6982e-06,SH14-023:1.0821e-06)1:1.0822e-06)1:4.5e-09)1:4.5e-09,(((SH14-002:5.3966e-06,SH14-015:1.6182e-06)1:5.379e-07,SH14-009:2.6982e-06)1:1.0821e-06,SH14-021:1.6182e-06)1:4.5e-09)1:4.5e-09,(((SH14-007:4.8568e-06,SH14-016:1.0821e-06)1:4.7e-09,SH14-019:1.0821e-06)1:1.0821e-06,SH14-014:3.2378e-06)1:1.23e-08)1:4.5e-09,SH14-028:4.5e-09)1:4.5e-09,(SH14-010:0.000133791,SH14-020:3.7775e-06)1:4.5e-09)1:4.5e-09,((SH14-008:3.2379e-06,SH14-025:3.7776e-06)1:1.0821e-06,(SH14-018:5.3965e-06,SH14-022:4.3171e-06)1:5.374e-07)1:4.5e-09)1:4.5e-09,(SH14-013:1.0821e-06,SH14-017:0.000133791)1:4.5e-09)1:4.5e-09,SH14-006:2.6982e-06)1:2.255e-07,SH14-027:2.4738e-06)1:7.5915e-06,SH14-012:0.00420588)1:0.000545308)1:5.1988e-06,SH10-30:0.000498294)1:0.00241787,((SH10-001:0.000407419,(SH11-001:0.000141943,SH12-011:0.000531153)1:1.0127e-06)1:4.4732e-06,((((SH12-001:7.5521e-06,SH12-010:0.000121911)1:4.5e-09,(((SH12-002:5.9026e-06,SH12-007:0.000125719)1:0.000123561,(SH12-004:3.2394e-06,SH12-005:7.0127e-06)1:1.0788e-06)1:4.5e-09,(SH12-003:4.3147e-06,SH12-009:2.6996e-06)1:4.5e-09)1:4.5e-09)1:4.5e-09,SH12-008:1.0788e-06)1:1.079e-06,SH12-006:5.3941e-06)1:0.000630848)1:0.00286185)1:0.000495688,((SH10-014:0.000243911,(SH10-015:3.607e-06,(SH11-002:0.000137025,((((SH13-001:7.5517e-06,SH13-005:4.8546e-06)1:4.5e-09,((SH13-003:8.6306e-06,SH13-007:1.6165e-06)1:1.0781e-06,(SH13-004:0.00014619,SH13-006:3.2343e-06)1:1.0767e-06)1:4.5e-09)1:4.5e-09,SH13-002:2.1553e-06)1:5.191e-07,SH13-008:0.000117613)1:0.000261718)1:0.000142052)1:7.6753e-06)1:0.00560301,((SH12-013:0.00231774,reference:4.5e-09)1:0.00367708,SH12-014:0.00358299)1:0.00138402)1:2.43533e-05)1:0.00214974,(SH10-002:0.00186723,SH12-012:0.00176948)1:1.36979e-05);"
     ]
    }
   ],
   "source": [
    "!cat thesis.core.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "postal-russian",
   "metadata": {},
   "source": [
    "## Profiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "surprised-thanksgiving",
   "metadata": {},
   "source": [
    "import pstats\n",
    "from pstats import SortKey\n",
    "p_tree = pstats.Stats('profile.tree.data').strip_dirs().sort_stats(SortKey.TIME, SortKey.NFL)\n",
    "# p_tree.print_stats(10)\n",
    "p_tree.print_stats(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "agreed-strain",
   "metadata": {},
   "source": [
    "# Database size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "stable-finish",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table                    Data size (MB)  Index size (MB)\n",
      "reference                0.02            0.00\n",
      "reference_sequence       0.02            0.02\n",
      "sample                   0.02            0.00\n",
      "sample_sequence          35.52           0.02\n",
      "sample_variation_allele  53.58           91.30\n",
      "variation_allele         10.55           5.52\n",
      "Total Data (MB)   99.71\n",
      "Total Index (MB)  96.86\n",
      "Total (MB)        196.57\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Fix warning about locale unset\n",
    "os.environ['LANG'] = 'en_US.UTF-8'\n",
    "\n",
    "!mysql -utest -ptest < query-db-size.txt | tee /tmp/total-db-sizes.txt | column -s$'\\t' -t\n",
    "!perl -e '$sd=0;$si=0;$s=0; while(<>){chomp; ($n,$ds,$is) = ($_ =~ /^(\\S+)\\s+(\\S+)\\s+(\\S+)/); if ($n ne \"Table\"){$s += $ds + $is; $si+=$is; $sd += $ds;}} print(\"\\nTotal Data (MB)\\t$sd\\nTotal Index (MB)\\t$si\\nTotal (MB)\\t$s\\n\");' < /tmp/total-db-sizes.txt \\\n",
    "    | column -s$'\\t' -t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "advanced-reach",
   "metadata": {},
   "source": [
    "# Compare above with size of input files\n",
    "\n",
    "This copies all the necessary input files to a temporary directory, compresses them, and checks size. This is to compare with size of storing information in database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "instrumental-strip",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total size\n",
      "86M\t/tmp/tmp2thoesqu\n",
      "Size of fasta\n",
      "86M\ttotal\n",
      "Size of VCF\n",
      "236K\ttotal\n",
      "Count of VCF\n",
      "59\n"
     ]
    }
   ],
   "source": [
    "import tempfile\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "with tempfile.TemporaryDirectory() as tmp_dir:\n",
    "    phy_path = os.path.abspath('phylogeny')\n",
    "    phy_names = [d for d in os.listdir(phy_path) if (os.path.isdir(f'{phy_path}/{d}') and d.startswith('SH'))]\n",
    "    fasta_files = []\n",
    "    for name in phy_names:\n",
    "        os.symlink(f'{phy_path}/{name}/snps.vcf.gz', f'{tmp_dir}/{name}.snps.vcf.gz')\n",
    "        shutil.copyfile(f'{phy_path}/{name}/snps.aligned.fa', f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "        fasta_files.append(f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "    !parallel -j 32 -I% gzip % ::: {' '.join(fasta_files)}\n",
    "    print('Total size')\n",
    "    !du -sh {tmp_dir}\n",
    "    print('Size of fasta')\n",
    "    !du -ch {tmp_dir}/*.snps.aligned.fa.gz | tail -n 1\n",
    "    print('Size of VCF')\n",
    "    !du -ch {tmp_dir}/*.snps.vcf.gz | tail -n 1\n",
    "    print('Count of VCF')\n",
    "    !ls {tmp_dir}/*.snps.vcf.gz | wc -l"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
