{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "classified-sword",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "Commits of thesis-index code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "fuzzy-recognition",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/workspace/thesis-data-simulation/jackalope/salmonella\n",
      "313608bcdc8ba539e65b9ad0f7dc9fb509acc53a\n",
      "0.1.0.dev1\n"
     ]
    }
   ],
   "source": [
    "!pwd\n",
    "!git -C ../../thesis-index rev-parse HEAD\n",
    "from storage import __version__\n",
    "print(__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "brave-vegetarian",
   "metadata": {},
   "source": [
    "# Load up variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "social-marriage",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/.local/lib/python3.8/site-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.4) or chardet (3.0.4) doesn't match a supported version!\n",
      "  warnings.warn(\"urllib3 ({}) or chardet ({}) doesn't match a supported \"\n",
      "\u001b[32m2021-03-23 17:37:02\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,56:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis_after_refactor?charset=utf8mb4\n",
      "\u001b[32m2021-03-23 17:37:02\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,60:\u001b[0m Use database directory db-after-refactor\n",
      "Loading phylogeny\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/thesis-index-refactor-relational-db/lib/python3.8/site-packages/sqlalchemy/sql/compiler.py:362: SAWarning: SELECT statement has a cartesian product between FROM element(s) \"reference\" and FROM element \"sample\".  Apply join condition(s) between each element to resolve.\n",
      "  util.warn(message)\n",
      "Loaded variants from [phylogeny] into database\n",
      "Took 3.66 minutes\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import shutil\n",
    "import time\n",
    "\n",
    "start_time = time.time()\n",
    "!variants \\\n",
    "    --config config-after-bcftools-refactor.yaml --verbose \\\n",
    "    load-snippy --reference-file input/S_HeidelbergSL476.fasta.gz phylogeny\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "straight-stylus",
   "metadata": {},
   "source": [
    "Make sure data loaded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "canadian-compact",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/.local/lib/python3.8/site-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.4) or chardet (3.0.4) doesn't match a supported version!\n",
      "  warnings.warn(\"urllib3 ({}) or chardet ({}) doesn't match a supported \"\n",
      "S_HeidelbergSL476\n"
     ]
    }
   ],
   "source": [
    "!variants --config config-after-bcftools-refactor.yaml list --type genome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "exciting-sharing",
   "metadata": {},
   "source": [
    "# Build tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "third-bacon",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/CSCScience.ca/apetkau/.local/lib/python3.8/site-packages/requests/__init__.py:89: RequestsDependencyWarning: urllib3 (1.26.4) or chardet (3.0.4) doesn't match a supported version!\n",
      "  warnings.warn(\"urllib3 ({}) or chardet ({}) doesn't match a supported \"\n",
      "\u001b[32m2021-03-23 17:40:44\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,56:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis_after_refactor?charset=utf8mb4\n",
      "\u001b[32m2021-03-23 17:40:44\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,60:\u001b[0m Use database directory db-after-refactor\n",
      "/home/CSCScience.ca/apetkau/miniconda3/envs/thesis-index-refactor-relational-db/lib/python3.8/site-packages/sqlalchemy/sql/compiler.py:362: SAWarning: SELECT statement has a cartesian product between FROM element(s) \"reference\" and FROM element \"sample\".  Apply join condition(s) between each element to resolve.\n",
      "  util.warn(message)\n",
      "\u001b[32m2021-03-23 17:40:44\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService.construct_alignment,91:\u001b[0m \u001b[32mStarted building alignment for 59 samples\u001b[0m\n",
      "\u001b[32m2021-03-23 17:40:53\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.CoreAlignmentService.construct_alignment,159:\u001b[0m \u001b[32mFinished building alignment for 59 samples. Took 9.22 seconds\u001b[0m\n",
      "\u001b[32m2021-03-23 17:40:55\u001b[0m \u001b[1;30mDEBUG\u001b[0m \u001b[34mstorage.variant.service.TreeService.build_tree,77:\u001b[0m \u001b[32mRunning: iqtree --terrace -T 52 -s /tmp/tmpyu191q9o/input.fasta -m MFP\u001b[0m\n",
      "Wrote tree to [thesis.full.tre]\n",
      "Wrote log file to [thesis.full.tre.log]\n",
      "Took 10.96 minutes\n"
     ]
    }
   ],
   "source": [
    "start_time = time.time()\n",
    "!variants \\\n",
    "    --config config-after-bcftools-refactor.yaml --verbose \\\n",
    "    tree --reference-name S_HeidelbergSL476 --output-file thesis.full.tre --align-type full --threads 52\n",
    "end_time = time.time()\n",
    "print(f'Took {(end_time - start_time)/60:0.2f} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "vietnamese-world",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(SH08-001:0.000746112,((((SH09-29:7.99266e-05,((((((((((SH14-001:3.42528e-05,SH14-010:4.08071e-05)1:1.531e-07,(((SH14-007:1.4062e-06,SH14-016:3.411e-07)1:3.412e-07,SH14-019:1.853e-07)1:3.412e-07,SH14-014:8.276e-07)1:1.531e-07)1:1.531e-07,(SH14-013:3.411e-07,SH14-020:1.0196e-06)1:1.531e-07)1:1.531e-07,((SH14-018:1.5653e-06,SH14-022:1.1678e-06)1:2.913e-07,SH14-024:3.411e-07)1:1.531e-07)1:1.531e-07,(SH14-006:7.151e-07,SH14-017:4.08077e-05)1:1.531e-07)1:1.531e-07,(((SH14-002:1.5652e-06,SH14-015:4.808e-07)1:3.886e-07,SH14-009:7.673e-07)1:3.412e-07,SH14-027:6.817e-07)1:1.531e-07)1:1.531e-07,((SH14-003:6.684e-07,SH14-026:9.194e-07)1:1.531e-07,SH14-028:1.531e-07)1:1.531e-07)1:1.531e-07,((SH14-004:2.1743e-06,SH14-011:1.9018e-06)1:1.531e-07,((SH14-005:1.0335e-06,SH14-023:1.844e-07)1:3.412e-07,(SH14-008:8.967e-07,SH14-025:1.0077e-06)1:1.531e-07)1:3.411e-07)1:1.531e-07)1:1.531e-07,SH14-021:4.809e-07)1:2.2572e-06,SH14-012:0.00128268)1:0.000166372)1:1.4799e-06,SH10-30:0.000152028)1:0.000737539,((SH10-001:0.000124319,(SH11-001:4.3197e-05,SH12-011:0.000162073)1:3.292e-07)1:1.3564e-06,((((((SH12-001:2.2017e-06,SH12-010:3.72023e-05)1:1.531e-07,(SH12-008:3.41e-07,SH12-009:7.575e-07)1:1.531e-07)1:1.531e-07,(SH12-003:1.1546e-06,SH12-006:1.897e-06)1:1.531e-07)1:1.531e-07,(SH12-002:1.7471e-06,SH12-007:3.8261e-05)1:3.77008e-05)1:3.37e-07,SH12-005:2.0576e-06)1:4.478e-07,SH12-004:4.522e-07)1:0.000192555)1:0.000872964)1:0.000151358,((SH10-014:7.43881e-05,(SH10-015:9.437e-07,(SH11-002:4.18503e-05,((((SH13-001:2.2235e-06,(SH13-004:4.46494e-05,SH13-006:9.194e-07)1:3.412e-07)1:1.531e-07,(SH13-002:6.683e-07,(SH13-003:2.5151e-06,SH13-007:4.809e-07)1:3.003e-07)1:1.531e-07)1:1.531e-07,SH13-005:1.4062e-06)1:2.12e-07,SH13-008:3.58272e-05)1:7.98273e-05)1:4.33933e-05)1:2.3317e-06)1:0.0017107,((SH12-013:0.000707626,reference:3.2e-07)1:0.00112212,SH12-014:0.00109331)1:0.000422634)1:7.3412e-06)1:0.000655984,(SH10-002:0.000569893,SH12-012:0.000540109)1:4.0093e-06);"
     ]
    }
   ],
   "source": [
    "!cat thesis.core.tre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "little-consciousness",
   "metadata": {},
   "source": [
    "# Database size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "affected-ensemble",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mysql: [Warning] Using a password on the command line interface can be insecure.\n",
      "Table                        Data size (MB)  Index size (MB)\n",
      "nucleotide_variants_samples  9.52            0.00\n",
      "reference                    0.02            0.00\n",
      "reference_sequence           0.02            0.02\n",
      "sample                       0.02            0.00\n",
      "sample_kmer_index            0.02            0.00\n",
      "sample_nucleotide_variation  0.02            0.02\n",
      "\n",
      "Total Data (MB)   9.62\n",
      "Total Index (MB)  0.04\n",
      "Total (MB)        9.66\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Fix warning about locale unset\n",
    "os.environ['LANG'] = 'en_US.UTF-8'\n",
    "\n",
    "tmp_out = '/tmp/total-db-sizes-after.txt'\n",
    "db_name = 'thesis_after_refactor'\n",
    "db_dir = 'db-after-refactor'\n",
    "query_file = '/tmp/query-after.txt'\n",
    "\n",
    "db_size_query = f'SELECT table_name AS `Table`,round(((data_length) / 1024 / 1024), 2) as `Data size (MB)`, round(((index_length) / 1024 / 1024), 2) as `Index size (MB)` FROM information_schema.TABLES WHERE table_schema = \"{db_name}\";'\n",
    "with open(query_file, 'w') as f:\n",
    "    f.write(db_size_query)\n",
    "!mysql -utest -ptest < {query_file} | tee {tmp_out} | column -s$'\\t' -t\n",
    "!echo\n",
    "!cat /tmp/total-db-sizes-after.txt | perl -e '$sd=0;$si=0;$s=0; while(<>){chomp; ($n,$ds,$is) = ($_ =~ /^(\\S+)\\s+(\\S+)\\s+(\\S+)/); if ($n ne \"Table\"){$s += $ds + $is; $si+=$is; $sd += $ds;}} print(\"\\nTotal Data (MB)\\t$sd\\nTotal Index (MB)\\t$si\\nTotal (MB)\\t$s\\n\");' | column -s$'\\t' -t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "deluxe-islam",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27M\tdb-after-refactor\n"
     ]
    }
   ],
   "source": [
    "!du -sh db-after-refactor"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "collective-optimum",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mysql: [Warning] Using a password on the command line interface can be insecure.\n",
      "count(*)\n",
      "76534\n"
     ]
    }
   ],
   "source": [
    "!echo \"select count(*) from nucleotide_variants_samples;\" | mysql -utest -ptest --database {db_name}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "floppy-feeding",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mysql: [Warning] Using a password on the command line interface can be insecure.\n",
      "count(*)\n",
      "59\n"
     ]
    }
   ],
   "source": [
    "!echo \"select count(*) from sample_nucleotide_variation;\" | mysql -utest -ptest --database {db_name}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "corrected-vegetarian",
   "metadata": {},
   "source": [
    "# Compare above with size of input files\n",
    "\n",
    "This copies all the necessary input files to a temporary directory, compresses them, and checks size. This is to compare with size of storing information in database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "documentary-cartridge",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total size\n",
      "111M\t/tmp/tmp9vjz6mmp\n",
      "Size of fasta\n",
      "86M\ttotal\n",
      "Size of VCF\n",
      "26M\ttotal\n",
      "Count of VCF\n",
      "59\n",
      "Size of fasta reference\n",
      "1.5M\tinput/S_HeidelbergSL476.fasta.gz\n"
     ]
    }
   ],
   "source": [
    "import tempfile\n",
    "import os\n",
    "import shutil\n",
    "\n",
    "with tempfile.TemporaryDirectory() as tmp_dir:\n",
    "    phy_path = os.path.abspath('phylogeny')\n",
    "    phy_names = [d for d in os.listdir(phy_path) if (os.path.isdir(f'{phy_path}/{d}') and d.startswith('SH'))]\n",
    "    fasta_files = []\n",
    "    for name in phy_names:\n",
    "        os.symlink(f'{phy_path}/{name}/snps.vcf.gz', f'{tmp_dir}/{name}.snps.vcf.gz')\n",
    "        shutil.copyfile(f'{phy_path}/{name}/snps.aligned.fa', f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "        fasta_files.append(f'{tmp_dir}/{name}.snps.aligned.fa')\n",
    "    !parallel -j 32 -I% gzip % ::: {' '.join(fasta_files)}\n",
    "    print('Total size')\n",
    "    !du -Lsh {tmp_dir}\n",
    "    print('Size of fasta')\n",
    "    !du -Lch {tmp_dir}/*.snps.aligned.fa.gz | tail -n 1\n",
    "    print('Size of VCF')\n",
    "    !du -Lch {tmp_dir}/*.snps.vcf.gz | tail -n 1\n",
    "    print('Count of VCF')\n",
    "    !ls {tmp_dir}/*.snps.vcf.gz | wc -l\n",
    "    print('Size of fasta reference')\n",
    "    !du -h input/S_HeidelbergSL476.fasta.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "prompt-guatemala",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
