{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "social-birmingham",
   "metadata": {},
   "source": [
    "# Measure performance for different methods of querying variant data\n",
    "\n",
    "First let's print out the exact git commit of my thesis-index code I am using."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "vanilla-booking",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41a96d85dbbb3b5910206a5fff4455dcb1cb1785\n"
     ]
    }
   ],
   "source": [
    "!git -C ../../../thesis-index rev-parse HEAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "monetary-privilege",
   "metadata": {},
   "outputs": [],
   "source": [
    "from storage.variant.io.SnippyVariantsReader import SnippyVariantsReader\n",
    "from pathlib import Path\n",
    "from os import listdir, path\n",
    "\n",
    "snippy_dir = Path('..', 'phylogeny')\n",
    "sample_dirs = [snippy_dir / d for d in listdir(snippy_dir) if path.isdir(snippy_dir / d)]\n",
    "\n",
    "variants_reader = SnippyVariantsReader(sample_dirs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "educational-volume",
   "metadata": {},
   "source": [
    "A function used to wrap around another function and measure runtime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "advisory-parallel",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "from statistics import mean, stdev\n",
    "import pandas as pd\n",
    "\n",
    "def get_runtime(func, **args):\n",
    "    start = time.time()\n",
    "    value = func(**args)\n",
    "    end = time.time()\n",
    "    runtime = end - start\n",
    "    return value, end - start\n",
    "\n",
    "def runtime_iteration(func, table: pd.DataFrame = None, name: str = None, num_samples: int = None, iterations = 5, print_result=True, **args):\n",
    "    runtimes = []\n",
    "    results = []\n",
    "    for i in range(0, iterations):\n",
    "        result, runtime = get_runtime(func, **args)\n",
    "        runtimes.append(runtime)\n",
    "        results.append(result)\n",
    "    \n",
    "    m_runtime = mean(runtimes)\n",
    "    s_runtime = stdev(runtimes) if len(runtimes) > 1 else 0\n",
    "    \n",
    "    if print_result:\n",
    "        result_value = results[0]\n",
    "        print(f'Runtime (m+-s): {m_runtime:0.2f} +- {s_runtime:0.2f} seconds, iters {iterations}, result {set(results)}')\n",
    "    else:\n",
    "        result_value = pd.NA\n",
    "        print(f'Runtime (m+-s): {m_runtime:0.2f} +- {s_runtime:0.2f} seconds, iters {iterations}')\n",
    "        \n",
    "    if table is not None:\n",
    "        # Drop previously-ran matching results so I don't have duplicate cases\n",
    "        drop_index = table[(table['Case'] == name) & (table['Number of Samples'] == num_samples)].index\n",
    "        table = table.drop(drop_index, axis='index')\n",
    "\n",
    "        table = table.append({\n",
    "            'Case': name,\n",
    "            'Number of Samples': num_samples,\n",
    "            'Iterations': iterations,\n",
    "            'Mean runtime (s)': m_runtime,\n",
    "            'Stdev runtime (s)': s_runtime,\n",
    "            'Result': result_value\n",
    "        }, ignore_index=True)\n",
    "              \n",
    "    return table, results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "subtle-recruitment",
   "metadata": {},
   "source": [
    "# Test variants union\n",
    "\n",
    "Let's test getting set of all variants among a list of samples (union)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "tropical-swing",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>Number of Samples</th>\n",
       "      <th>Iterations</th>\n",
       "      <th>Mean runtime (s)</th>\n",
       "      <th>Stdev runtime (s)</th>\n",
       "      <th>Result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Case, Number of Samples, Iterations, Mean runtime (s), Stdev runtime (s), Result]\n",
       "Index: []"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "case1_samples = ['SH12-001']\n",
    "case2_samples = [\"SH14-004\" ,\"SH13-001\" ,\"SH14-011\" ,\"SH14-016\" ,\"SH09-29\" ,\"SH12-008\" ,\"SH14-010\" ,\"SH14-028\" ,\"SH10-30\" ,\"SH12-007\"]\n",
    "\n",
    "results_table = pd.DataFrame(data=[], columns=[\n",
    "    'Case',\n",
    "    'Number of Samples',\n",
    "    'Iterations',\n",
    "    'Mean runtime (s)',\n",
    "    'Stdev runtime (s)',\n",
    "    'Result',\n",
    "])\n",
    "results_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "raised-cement",
   "metadata": {},
   "source": [
    "## 1. From VCF files\n",
    "\n",
    "### 1.1. VCF including loading file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "warming-appreciation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime (m+-s): 1.14 +- 0.03 seconds, iters 5, result {17193}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime (m+-s): 11.91 +- 0.06 seconds, iters 5, result {36920}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>Number of Samples</th>\n",
       "      <th>Iterations</th>\n",
       "      <th>Mean runtime (s)</th>\n",
       "      <th>Stdev runtime (s)</th>\n",
       "      <th>Result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Load VCF as DataFrame</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.135049</td>\n",
       "      <td>0.030889</td>\n",
       "      <td>17193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Load VCF as DataFrame</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>11.914648</td>\n",
       "      <td>0.061704</td>\n",
       "      <td>36920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Case Number of Samples Iterations  Mean runtime (s)  \\\n",
       "0  Load VCF as DataFrame                 1          5          1.135049   \n",
       "1  Load VCF as DataFrame                10          5         11.914648   \n",
       "\n",
       "   Stdev runtime (s) Result  \n",
       "0           0.030889  17193  \n",
       "1           0.061704  36920  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from typing import List, Set\n",
    "\n",
    "def get_union_from_vcf(samples: List[str]) -> int:\n",
    "    sample_dirs = [snippy_dir / s for s in samples]\n",
    "    variants_reader = SnippyVariantsReader(sample_dirs)\n",
    "    var_df = variants_reader.get_variants_table()\n",
    "    \n",
    "    vars_union = set()\n",
    "    var_df['SPDI'] = var_df['CHROM'] + ':' + var_df['POS'].astype(str) + ':' + var_df['REF'] + ':' + var_df['ALT']\n",
    "    for sample in samples:\n",
    "        vars_union = vars_union.union(set(var_df[var_df['SAMPLE'] == sample]['SPDI'].tolist()))\n",
    "        \n",
    "    return len(vars_union)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "results_table, results = runtime_iteration(get_union_from_vcf, results_table, 'Load VCF as DataFrame', len(case1_samples), samples=case1_samples)\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "results_table, results = runtime_iteration(get_union_from_vcf, results_table, 'Load VCF as DataFrame', len(case2_samples), samples=case2_samples)\n",
    "results_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deadly-recipe",
   "metadata": {},
   "source": [
    "### 1.2. VCF without loading file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "recorded-return",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime (m+-s): 0.02 +- 0.00 seconds, iters 5, result {17193}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime (m+-s): 0.36 +- 0.02 seconds, iters 5, result {36920}\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>Number of Samples</th>\n",
       "      <th>Iterations</th>\n",
       "      <th>Mean runtime (s)</th>\n",
       "      <th>Stdev runtime (s)</th>\n",
       "      <th>Result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Load VCF as DataFrame</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1.135049</td>\n",
       "      <td>0.030889</td>\n",
       "      <td>17193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Load VCF as DataFrame</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>11.914648</td>\n",
       "      <td>0.061704</td>\n",
       "      <td>36920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Load VCF as DataFrame (no file load)</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.024153</td>\n",
       "      <td>0.002352</td>\n",
       "      <td>17193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Load VCF as DataFrame (no file load)</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.363590</td>\n",
       "      <td>0.022145</td>\n",
       "      <td>36920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Case Number of Samples Iterations  \\\n",
       "0                 Load VCF as DataFrame                 1          5   \n",
       "1                 Load VCF as DataFrame                10          5   \n",
       "2  Load VCF as DataFrame (no file load)                 1          5   \n",
       "3  Load VCF as DataFrame (no file load)                10          5   \n",
       "\n",
       "   Mean runtime (s)  Stdev runtime (s) Result  \n",
       "0          1.135049           0.030889  17193  \n",
       "1         11.914648           0.061704  36920  \n",
       "2          0.024153           0.002352  17193  \n",
       "3          0.363590           0.022145  36920  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def get_union_from_vcf_no_load(var_df: pd.DataFrame) -> int:\n",
    "    vars_union = set()\n",
    "    samples = set(var_df['SAMPLE'].tolist())\n",
    "    var_df['SPDI'] = var_df['CHROM'] + ':' + var_df['POS'].astype(str) + ':' + var_df['REF'] + ':' + var_df['ALT']\n",
    "    for sample in samples:\n",
    "        vars_union = vars_union.union(set(var_df[var_df['SAMPLE'] == sample]['SPDI'].tolist()))\n",
    "        \n",
    "    return len(vars_union)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "\n",
    "sample_dirs = [snippy_dir / s for s in case1_samples]\n",
    "variants_reader = SnippyVariantsReader(sample_dirs)\n",
    "var_df = variants_reader.get_variants_table()\n",
    "results_table, results = runtime_iteration(get_union_from_vcf_no_load, results_table, 'Load VCF as DataFrame (no file load)', len(case1_samples), var_df=var_df)\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "\n",
    "sample_dirs = [snippy_dir / s for s in case2_samples]\n",
    "variants_reader = SnippyVariantsReader(sample_dirs)\n",
    "var_df = variants_reader.get_variants_table()\n",
    "results_table, results = runtime_iteration(get_union_from_vcf_no_load, results_table, 'Load VCF as DataFrame (no file load)', len(case2_samples), var_df=var_df)\n",
    "results_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "necessary-branch",
   "metadata": {},
   "source": [
    "## 2. From relational database\n",
    "\n",
    "To test this out, I first have to load all the VCF files into the database, which takes a while. Since I'm only testing a specific query (finding union of all variants in some samples) I'm not timing this loading time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "improving-spirituality",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-02-24 13:50:10\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,53:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-02-24 13:50:10\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,56:\u001b[0m Use seqrepo directory seq_repo\n",
      "Loading ../phylogeny\n",
      "\u001b[32m2021-02-24 13:50:10\u001b[0m \u001b[1;30mWARNING\u001b[0m \u001b[34mstorage.load_variants_common,90:\u001b[0m \u001b[33mReference genome [../input/S_HeidelbergSL476.fasta.gz] already exists, will not load\u001b[0m\n",
      "\u001b[32m2021-02-24 13:50:10\u001b[0m \u001b[1;30mERROR\u001b[0m \u001b[34mstorage.load_variants_common,94:\u001b[0m \u001b[31mSamples ['SH14-013', 'SH14-001', 'SH13-006', 'SH12-009', 'SH12-011', 'SH12-004', 'SH14-008', 'SH14-026', 'SH12-007', 'SH13-005', 'SH14-018', 'SH13-003', 'SH13-002', 'SH14-020', 'SH14-014', 'SH11-001', 'SH14-011', 'SH14-002', 'SH12-005', 'SH13-007', 'SH14-024', 'SH13-004', 'SH14-012', 'SH10-015', 'SH13-008', 'SH14-006', 'SH14-023', 'SH10-014', 'SH14-009', 'SH12-008', 'SH12-010', 'SH12-013', 'SH10-002', 'SH12-003', 'SH14-004', 'SH14-025', 'SH14-019', 'SH14-007', 'SH10-30', 'SH14-017', 'SH12-014', 'SH14-027', 'SH12-002', 'SH12-001', 'SH14-003', 'SH08-001', 'SH14-010', 'SH14-021', 'SH14-016', 'SH11-002', 'SH09-29', 'SH14-028', 'SH14-005', 'SH12-006', 'SH13-001', 'SH14-015', 'SH14-022', 'SH12-012', 'SH10-001'] already exist, will not load any variants\u001b[0m\n",
      "Took 0.0 minutes\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "!variants --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --seqrepo-dir seq_repo --verbose \\\n",
    "    load-snippy --reference-file ../input/S_HeidelbergSL476.fasta.gz ../phylogeny\n",
    "end = time.time()\n",
    "print(f'Took {(end-start)/60:0.1f} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "controlling-party",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "from sqlalchemy.orm import sessionmaker\n",
    "\n",
    "def create_session():\n",
    "    engine = create_engine('mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4', echo=False)\n",
    "\n",
    "    Session = sessionmaker(bind=engine)\n",
    "    session = Session()\n",
    "    return session"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "hawaiian-worry",
   "metadata": {},
   "source": [
    "### 2.1. No checking for proper reference/sequence name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "fourth-straight",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime (m+-s): 0.42 +- 0.04 seconds, iters 5, result {17193}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime (m+-s): 3.24 +- 0.06 seconds, iters 5, result {36920}\n"
     ]
    }
   ],
   "source": [
    "from storage.variant.model import Sample\n",
    "\n",
    "def get_union_from_relational_db(samples: List[str]) -> int:\n",
    "    session = create_session()\n",
    "    sample_objs = session.query(Sample).filter(Sample.name.in_(samples)).all()\n",
    "    \n",
    "    vars_union = set()\n",
    "    for sample in sample_objs:\n",
    "        svars = {v.id for v in sample.variants}\n",
    "        vars_union = vars_union.union(svars)\n",
    "        \n",
    "    return len(vars_union)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "results_table, results = runtime_iteration(get_union_from_relational_db, results_table, 'Relational DB (no check proper ref)', len(case1_samples), samples=case1_samples)\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "results_table, results = runtime_iteration(get_union_from_relational_db, results_table, 'Relational DB (no check proper ref)', len(case2_samples), samples=case2_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "married-australian",
   "metadata": {},
   "source": [
    "### 2.2. Checking for proper reference/sequence name from database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "early-brighton",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime (m+-s): 2.19 +- 0.03 seconds, iters 5, result {17193}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime (m+-s): 3.06 +- 0.14 seconds, iters 5, result {36920}\n"
     ]
    }
   ],
   "source": [
    "from storage.variant.model import VariationAllele, ReferenceSequence, Reference\n",
    "\n",
    "def get_union_from_relational_db_via_query(samples: List[str]) -> int:\n",
    "    session = create_session()\n",
    "    variants = session.query(VariationAllele) \\\n",
    "        .select_from(Sample) \\\n",
    "        .join(Sample.variants) \\\n",
    "        .join(ReferenceSequence) \\\n",
    "        .join(Reference) \\\n",
    "        .filter(Reference.name == 'S_HeidelbergSL476') \\\n",
    "        .filter(Sample.name.in_(samples)) \\\n",
    "        .all()\n",
    "    \n",
    "    vars_union = {v.id for v in variants}\n",
    "    return len(vars_union)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "results_table, results = runtime_iteration(get_union_from_relational_db_via_query, results_table, 'Relational DB (proper ref as query)', len(case1_samples), samples=case1_samples)\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "results_table, results = runtime_iteration(get_union_from_relational_db_via_query, results_table, 'Relational DB (proper ref as query)', len(case2_samples), samples=case2_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "lucky-surge",
   "metadata": {},
   "source": [
    "### 2.3. Checking for proper reference/sequence name in code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "strong-juvenile",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime (m+-s): 0.41 +- 0.04 seconds, iters 5, result {17193}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime (m+-s): 3.30 +- 0.06 seconds, iters 5, result {36920}\n"
     ]
    }
   ],
   "source": [
    "from storage.variant.model import Sample, Reference, ReferenceSequence\n",
    "\n",
    "def get_union_from_relational_db_via_code(samples: List[str]) -> int:\n",
    "    session = create_session()\n",
    "    sample_objs = session.query(Sample).filter(Sample.name.in_(samples)).all()\n",
    "    ref_sequences = session.query(ReferenceSequence) \\\n",
    "        .join(Reference) \\\n",
    "        .filter(Reference.name == 'S_HeidelbergSL476') \\\n",
    "        .all()\n",
    "    \n",
    "    ref_sequence_ids = {r.id for r in ref_sequences}\n",
    "    \n",
    "    vars_union = set()\n",
    "    for sample in sample_objs:\n",
    "        svars = {v.id for v in sample.variants if v.sequence_id in ref_sequence_ids}\n",
    "        vars_union = vars_union.union(svars)\n",
    "        \n",
    "    return len(vars_union)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "results_table, results = runtime_iteration(get_union_from_relational_db_via_code, results_table, 'Relational DB (proper ref as code)', len(case1_samples), samples=case1_samples)\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "results_table, results = runtime_iteration(get_union_from_relational_db_via_code, results_table, 'Relational DB (proper ref as code)', len(case2_samples), samples=case2_samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "established-executive",
   "metadata": {},
   "source": [
    "## 3. From Bloom filter\n",
    "\n",
    "I am testing out Bloom filters, but I note that I don't think there's any way to extract out the specific set of variants within a Bloom filter afterwards. At least not without storing a separate mapping of variant identifiers to the specific hashes used by the Bloom filters.\n",
    "\n",
    "Nor do I think I can even get the number of elements in a Bloom filter.\n",
    "\n",
    "### 3.1. In-memory Bloom filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "awful-reverse",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Building Bloom Filters\n",
      "Runtime (m+-s): 1.57 +- 0.07 seconds, iters 5\n",
      "Using Bloom Filters\n",
      "Runtime (m+-s): 0.00 +- 0.00 seconds, iters 5, result {'BloomFilter(ideal_num_elements_n=1000000, error_rate_p=0.100000, num_bits_m=4792530)'}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Building Bloom Filters\n",
      "Runtime (m+-s): 15.95 +- 0.05 seconds, iters 5\n",
      "Using Bloom Filters\n",
      "Runtime (m+-s): 0.30 +- 0.01 seconds, iters 5, result {'BloomFilter(ideal_num_elements_n=1000000, error_rate_p=0.100000, num_bits_m=4792530)'}\n"
     ]
    }
   ],
   "source": [
    "from typing import Dict\n",
    "import copy\n",
    "\n",
    "from bloom_filter import BloomFilter\n",
    "\n",
    "def load_vcf_to_bloom_filters(samples: List[str]) -> Dict[str, BloomFilter]:\n",
    "    sample_dirs = [snippy_dir / s for s in samples]\n",
    "    variants_reader = SnippyVariantsReader(sample_dirs)\n",
    "    var_df = variants_reader.get_variants_table()\n",
    "    \n",
    "    sample_filters = {}\n",
    "    \n",
    "    vars_union = set()\n",
    "    var_df['SPDI'] = var_df['CHROM'] + ':' + var_df['POS'].astype(str) + ':' + var_df['REF'] + ':' + var_df['ALT']\n",
    "    for sample in samples:\n",
    "        sample_filters[sample] = BloomFilter(max_elements=10**6, error_rate=0.1)\n",
    "        for v in var_df[var_df['SAMPLE'] == sample]['SPDI'].tolist():\n",
    "            sample_filters[sample].add(v)\n",
    "        \n",
    "    return sample_filters\n",
    "\n",
    "def union_from_bloom_filters(sample_filters: Dict[str, BloomFilter]) -> int:\n",
    "    samples = list(sample_filters.keys())\n",
    "    sample = samples.pop()\n",
    "    start_bloom = copy.deepcopy(sample_filters[sample])\n",
    "    for sample in samples:\n",
    "        start_bloom.union(sample_filters[sample])\n",
    "        \n",
    "    return repr(start_bloom)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "print('Building Bloom Filters')\n",
    "no_table, results_case1 = runtime_iteration(load_vcf_to_bloom_filters, print_result=False, samples=case1_samples)\n",
    "print('Using Bloom Filters')\n",
    "results_table, results = runtime_iteration(union_from_bloom_filters, results_table, 'Bloom filter (union)', len(case1_samples), sample_filters=results_case1[0])\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "print('Building Bloom Filters')\n",
    "no_table, results_case2 = runtime_iteration(load_vcf_to_bloom_filters, print_result=False, samples=case2_samples)\n",
    "print('Using Bloom Filters')\n",
    "results_table, results = runtime_iteration(union_from_bloom_filters, results_table, 'Bloom filter (union)', len(case2_samples), sample_filters=results_case2[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "irish-jewel",
   "metadata": {},
   "source": [
    "## 4. BCFTools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "afraid-cleanup",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Converting to BCF\n",
      "Runtime (m+-s): 0.40 +- 0.01 seconds, iters 5\n",
      "Finding union\n",
      "Runtime (m+-s): 0.16 +- 0.01 seconds, iters 5, result {17193}\n",
      "\n",
      "Case 2: 10 samples\n",
      "Converting to BCF\n",
      "Runtime (m+-s): 3.98 +- 0.04 seconds, iters 5\n",
      "Finding union\n",
      "Runtime (m+-s): 0.30 +- 0.01 seconds, iters 5, result {36920}\n"
     ]
    }
   ],
   "source": [
    "import tempfile\n",
    "from pathlib import Path\n",
    "import subprocess\n",
    "from io import StringIO\n",
    "\n",
    "def load_vcf_to_bcf(samples: List[str]) -> Dict[str, Path]:\n",
    "    bcf_dir = Path(tempfile.mkdtemp())\n",
    "    vcf_files = {s: snippy_dir / s / 'snps.vcf.gz' for s in samples}\n",
    "    bcf_files = {s: bcf_dir / f'{s}.bcf' for s in samples}\n",
    "    \n",
    "    for sample in vcf_files:\n",
    "        vcf_file = vcf_files[sample]\n",
    "        bcf_file = bcf_files[sample]\n",
    "        command_bcf = ['bcftools', 'view', str(vcf_file), '-o', str(bcf_file), '-O', 'b', '-l', '9']\n",
    "        command_index = ['bcftools', 'index', str(bcf_file)]\n",
    "        try:\n",
    "            subprocess.run(command_bcf, stdout=subprocess.PIPE, stderr=subprocess.PIPE,\n",
    "                                       check=True, text=True)\n",
    "            subprocess.run(command_index, stdout=subprocess.PIPE, stderr=subprocess.PIPE,\n",
    "                                       check=True, text=True)\n",
    "        except subprocess.CalledProcessError as e:\n",
    "            err_msg = str(e.stderr.strip())\n",
    "            raise Exception(f'Could not run bcftools on vcf_file=[{vcf_file}]: error {err_msg}')\n",
    "            \n",
    "    return bcf_files\n",
    "\n",
    "def union_from_bcf_files(sample_files: Dict[str, Path]) -> int:\n",
    "    command_isec = ['bcftools', 'isec', '-c', 'none', '-n', '+1', '--threads', '1']\n",
    "    for sample in sample_files:\n",
    "        command_isec.append(str(sample_files[sample]))\n",
    "        \n",
    "    command_single = None\n",
    "    command_single_cut = None\n",
    "    if len(sample_files) == 1:\n",
    "        sample = list(sample_files.keys())[0]\n",
    "        command_single = ['bcftools', 'view', '--threads', '1','--no-header', str(sample_files[sample])]\n",
    "        command_single_cut = ['cut', '-f', '1,2,4,5']\n",
    "        \n",
    "    try:\n",
    "        if command_single is not None:\n",
    "            result_first = subprocess.Popen(command_single, stdout=subprocess.PIPE, text=True)\n",
    "            result = subprocess.run(command_single_cut, stdin=result_first.stdout, stdout=subprocess.PIPE, text=True)\n",
    "        else:\n",
    "            result = subprocess.run(command_isec, stdout=subprocess.PIPE, stderr=subprocess.PIPE,\n",
    "                                   check=True, text=True)\n",
    "        vars_data = StringIO(result.stdout)\n",
    "        var_df = pd.read_csv(vars_data, sep='\\t', names=['CHROM', 'POS', 'REF', 'ALT', 'FILES'])\n",
    "        var_df['SPDI'] = var_df['CHROM'] + ':' + var_df['POS'].astype(str) + ':' + var_df['REF'] + ':' + var_df['ALT']\n",
    "        var_set = set(var_df['SPDI'].tolist())\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        err_msg = str(e.stderr.strip())\n",
    "        raise Exception(f'Could not run [{\" \".join(command)}]: error {err_msg}')\n",
    "        \n",
    "    return len(var_set)\n",
    "\n",
    "sam = [\"SH14-004\"]\n",
    "sample_bcf_files = load_vcf_to_bcf(sam)\n",
    "union_from_bcf_files(sample_bcf_files)\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "print('Converting to BCF')\n",
    "no_table, results_case1 = runtime_iteration(load_vcf_to_bcf, print_result=False, samples=case1_samples)\n",
    "print('Finding union')\n",
    "results_table, results = runtime_iteration(union_from_bcf_files, results_table, 'BCFTools', len(case1_samples), sample_files=results_case1[0])\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "print('Converting to BCF')\n",
    "no_table, results_case2 = runtime_iteration(load_vcf_to_bcf, print_result=False, samples=case2_samples)\n",
    "print('Finding union')\n",
    "results_table, results = runtime_iteration(union_from_bcf_files, results_table, 'BCFTools', len(case2_samples), sample_files=results_case2[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "preliminary-climate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Case</th>\n",
       "      <th>Number of Samples</th>\n",
       "      <th>Iterations</th>\n",
       "      <th>Mean runtime (s)</th>\n",
       "      <th>Stdev runtime (s)</th>\n",
       "      <th>Result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Relational DB (proper ref as query)</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>2.187969</td>\n",
       "      <td>0.025792</td>\n",
       "      <td>17193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Relational DB (proper ref as query)</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>3.062717</td>\n",
       "      <td>0.137326</td>\n",
       "      <td>36920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Relational DB (proper ref as code)</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.407293</td>\n",
       "      <td>0.036488</td>\n",
       "      <td>17193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Relational DB (proper ref as code)</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>3.297895</td>\n",
       "      <td>0.062367</td>\n",
       "      <td>36920</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bloom filter (union)</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.000238</td>\n",
       "      <td>0.000252</td>\n",
       "      <td>BloomFilter(ideal_num_elements_n=1000000, erro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bloom filter (union)</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.297692</td>\n",
       "      <td>0.005987</td>\n",
       "      <td>BloomFilter(ideal_num_elements_n=1000000, erro...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>BCFTools</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.162710</td>\n",
       "      <td>0.014770</td>\n",
       "      <td>17193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BCFTools</td>\n",
       "      <td>10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.304076</td>\n",
       "      <td>0.008903</td>\n",
       "      <td>36920</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  Case Number of Samples Iterations  \\\n",
       "0  Relational DB (proper ref as query)                 1          5   \n",
       "1  Relational DB (proper ref as query)                10          5   \n",
       "2   Relational DB (proper ref as code)                 1          5   \n",
       "3   Relational DB (proper ref as code)                10          5   \n",
       "4                 Bloom filter (union)                 1          5   \n",
       "5                 Bloom filter (union)                10          5   \n",
       "6                             BCFTools                 1          5   \n",
       "7                             BCFTools                10          5   \n",
       "\n",
       "   Mean runtime (s)  Stdev runtime (s)  \\\n",
       "0          2.187969           0.025792   \n",
       "1          3.062717           0.137326   \n",
       "2          0.407293           0.036488   \n",
       "3          3.297895           0.062367   \n",
       "4          0.000238           0.000252   \n",
       "5          0.297692           0.005987   \n",
       "6          0.162710           0.014770   \n",
       "7          0.304076           0.008903   \n",
       "\n",
       "                                              Result  \n",
       "0                                              17193  \n",
       "1                                              36920  \n",
       "2                                              17193  \n",
       "3                                              36920  \n",
       "4  BloomFilter(ideal_num_elements_n=1000000, erro...  \n",
       "5  BloomFilter(ideal_num_elements_n=1000000, erro...  \n",
       "6                                              17193  \n",
       "7                                              36920  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "institutional-termination",
   "metadata": {},
   "source": [
    "## 5. MongoDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "liable-photograph",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
