{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "similar-opera",
   "metadata": {},
   "source": [
    "# Measure performance for different methods of querying variant data\n",
    "\n",
    "First let's print out the exact git commit of my thesis-index code I am using."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "boolean-classification",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "41a96d85dbbb3b5910206a5fff4455dcb1cb1785\n"
     ]
    }
   ],
   "source": [
    "!git -C ../../../thesis-index rev-parse HEAD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "burning-vancouver",
   "metadata": {},
   "outputs": [],
   "source": [
    "from storage.variant.io.SnippyVariantsReader import SnippyVariantsReader\n",
    "from pathlib import Path\n",
    "from os import listdir, path\n",
    "\n",
    "snippy_dir = Path('..', 'phylogeny')\n",
    "sample_dirs = [snippy_dir / d for d in listdir(snippy_dir) if path.isdir(snippy_dir / d)]\n",
    "\n",
    "variants_reader = SnippyVariantsReader(sample_dirs)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "angry-scroll",
   "metadata": {},
   "source": [
    "A function used to wrap around another function and measure runtime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "theoretical-basic",
   "metadata": {},
   "outputs": [],
   "source": [
    "import time\n",
    "\n",
    "def get_runtime(func, **args):\n",
    "    start = time.time()\n",
    "    value = func(**args)\n",
    "    end = time.time()\n",
    "    runtime = end - start\n",
    "    print(f'Runtime: {runtime:0.2f} seconds')\n",
    "    return value, end - start"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "exotic-vegetation",
   "metadata": {},
   "source": [
    "# Test variants union\n",
    "\n",
    "Let's test getting set of all variants among a list of samples (union)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "blank-married",
   "metadata": {},
   "outputs": [],
   "source": [
    "case1_samples = ['SH12-001']\n",
    "case2_samples = [\"SH14-004\" ,\"SH13-001\" ,\"SH14-011\" ,\"SH14-016\" ,\"SH09-29\" ,\"SH12-008\" ,\"SH14-010\" ,\"SH14-028\" ,\"SH10-30\" ,\"SH12-007\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "contemporary-presence",
   "metadata": {},
   "source": [
    "## 1. From VCF files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "numeric-nerve",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime: 1.43 seconds\n",
      "Length of union: 17193\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime: 13.18 seconds\n",
      "Length of union: 36920\n"
     ]
    }
   ],
   "source": [
    "from typing import List, Set\n",
    "\n",
    "def get_union_from_vcf(samples: List[str]) -> Set[str]:\n",
    "    sample_dirs = [snippy_dir / s for s in samples]\n",
    "    variants_reader = SnippyVariantsReader(sample_dirs)\n",
    "    var_df = variants_reader.get_variants_table()\n",
    "    \n",
    "    vars_union = set()\n",
    "    var_df['SPDI'] = var_df['CHROM'] + ':' + var_df['POS'].astype(str) + ':' + var_df['REF'] + ':' + var_df['ALT']\n",
    "    for sample in samples:\n",
    "        vars_union = vars_union.union(set(var_df[var_df['SAMPLE'] == sample]['SPDI'].tolist()))\n",
    "        \n",
    "    return vars_union\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "union, runtime = get_runtime(get_union_from_vcf, samples=case1_samples)\n",
    "print(f'Length of union: {len(union)}')\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "union, runtime = get_runtime(get_union_from_vcf, samples=case2_samples)\n",
    "print(f'Length of union: {len(union)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "desperate-ontario",
   "metadata": {},
   "source": [
    "## 2. From relational database\n",
    "\n",
    "To test this out, I first have to load all the VCF files into the database, which takes a while. Since I'm only testing a specific query (finding union of all variants in some samples) I'm not timing this loading time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "informational-white",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[32m2021-02-23 15:32:07\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,53:\u001b[0m Connecting to database mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4\n",
      "\u001b[32m2021-02-23 15:32:07\u001b[0m \u001b[1;30mINFO\u001b[0m \u001b[34mstorage.main,56:\u001b[0m Use seqrepo directory seq_repo\n",
      "Loading ../phylogeny\n",
      "Loaded variants from [../phylogeny] into database\n",
      "Took 4.9 minutes\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "!variants --database-connection 'mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4' \\\n",
    "    --seqrepo-dir seq_repo --verbose \\\n",
    "    load-snippy --reference-file ../input/S_HeidelbergSL476.fasta.gz ../phylogeny\n",
    "end = time.time()\n",
    "print(f'Took {(end-start)/60:0.1f} minutes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "individual-illness",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import create_engine\n",
    "from sqlalchemy.orm import sessionmaker\n",
    "\n",
    "def create_session():\n",
    "    engine = create_engine('mysql+pymysql://test:test@localhost/thesis?charset=utf8mb4', echo=False)\n",
    "\n",
    "    Session = sessionmaker(bind=engine)\n",
    "    session = Session()\n",
    "    return session"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "superior-following",
   "metadata": {},
   "source": [
    "### 2.1. No checking for proper reference/sequence name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "identical-aviation",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime: 0.63 seconds\n",
      "Length of union: 17193\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime: 3.57 seconds\n",
      "Length of union: 36920\n"
     ]
    }
   ],
   "source": [
    "from storage.variant.model import Sample\n",
    "\n",
    "def get_union_from_relational_db(samples: List[str]) -> Set[str]:\n",
    "    session = create_session()\n",
    "    sample_objs = session.query(Sample).filter(Sample.name.in_(samples)).all()\n",
    "    \n",
    "    vars_union = set()\n",
    "    for sample in sample_objs:\n",
    "        svars = {v.id for v in sample.variants}\n",
    "        vars_union = vars_union.union(svars)\n",
    "        \n",
    "    return vars_union\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "union, runtime = get_runtime(get_union_from_relational_db, samples=case1_samples)\n",
    "print(f'Length of union: {len(union)}')\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "union, runtime = get_runtime(get_union_from_relational_db, samples=case2_samples)\n",
    "print(f'Length of union: {len(union)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fantastic-standing",
   "metadata": {},
   "source": [
    "### 2.2. Checking for proper reference/sequence name from database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "designed-watson",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime: 2.34 seconds\n",
      "Length of union: 17193\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime: 3.32 seconds\n",
      "Length of union: 36920\n"
     ]
    }
   ],
   "source": [
    "from storage.variant.model import VariationAllele, ReferenceSequence, Reference\n",
    "\n",
    "def get_union_from_relational_db_via_query(samples: List[str]) -> Set[str]:\n",
    "    session = create_session()\n",
    "    variants = session.query(VariationAllele) \\\n",
    "        .select_from(Sample) \\\n",
    "        .join(Sample.variants) \\\n",
    "        .join(ReferenceSequence) \\\n",
    "        .join(Reference) \\\n",
    "        .filter(Reference.name == 'S_HeidelbergSL476') \\\n",
    "        .filter(Sample.name.in_(samples)) \\\n",
    "        .all()\n",
    "    \n",
    "    vars_union = {v.id for v in variants}\n",
    "    return vars_union\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "union, runtime = get_runtime(get_union_from_relational_db_via_query, samples=case1_samples)\n",
    "print(f'Length of union: {len(union)}')\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "union, runtime = get_runtime(get_union_from_relational_db_via_query, samples=case2_samples)\n",
    "print(f'Length of union: {len(union)}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "improved-collectible",
   "metadata": {},
   "source": [
    "### 2.3. Checking for proper reference/sequence name in code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "employed-robinson",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case 1: Single (1) sample\n",
      "Runtime: 0.52 seconds\n",
      "Length of union: 17193\n",
      "\n",
      "Case 2: 10 samples\n",
      "Runtime: 3.31 seconds\n",
      "Length of union: 36920\n"
     ]
    }
   ],
   "source": [
    "from storage.variant.model import Sample, Reference, ReferenceSequence\n",
    "\n",
    "def get_union_from_relational_db_via_code(samples: List[str]) -> Set[str]:\n",
    "    session = create_session()\n",
    "    sample_objs = session.query(Sample).filter(Sample.name.in_(samples)).all()\n",
    "    ref_sequences = session.query(ReferenceSequence) \\\n",
    "        .join(Reference) \\\n",
    "        .filter(Reference.name == 'S_HeidelbergSL476') \\\n",
    "        .all()\n",
    "    \n",
    "    ref_sequence_ids = {r.id for r in ref_sequences}\n",
    "    \n",
    "    vars_union = set()\n",
    "    for sample in sample_objs:\n",
    "        svars = {v.id for v in sample.variants if v.sequence_id in ref_sequence_ids}\n",
    "        vars_union = vars_union.union(svars)\n",
    "        \n",
    "    return vars_union\n",
    "\n",
    "print(f'Case 1: Single ({len(case1_samples)}) sample')\n",
    "union, runtime = get_runtime(get_union_from_relational_db_via_code, samples=case1_samples)\n",
    "print(f'Length of union: {len(union)}')\n",
    "\n",
    "print(f'\\nCase 2: {len(case2_samples)} samples')\n",
    "union, runtime = get_runtime(get_union_from_relational_db_via_code, samples=case2_samples)\n",
    "print(f'Length of union: {len(union)}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
